<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>4 From disparity to macroevolution | Traits data in R</title>
  <meta name="description" content="A guide to using morphological trait data in R" />
  <meta name="generator" content="bookdown 0.36 and GitBook 2.6.7" />

  <meta property="og:title" content="4 From disparity to macroevolution | Traits data in R" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="A guide to using morphological trait data in R" />
  <meta name="github-repo" content="TGuillerme/Morphological_traits" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="4 From disparity to macroevolution | Traits data in R" />
  
  <meta name="twitter:description" content="A guide to using morphological trait data in R" />
  

<meta name="author" content="Thomas Guillerme (t.guillerme@sheffield.ac.uk)" />


<meta name="date" content="2024-08-08" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="from-traitspaces-to-disparity.html"/>

<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preamble</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#mega-disclaimer"><i class="fa fa-check"></i><b>1.1</b> Mega-disclaimer</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#requirements"><i class="fa fa-check"></i><b>1.2</b> Requirements</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#general-computer-level"><i class="fa fa-check"></i><b>1.2.1</b> General computer level</a></li>
<li class="chapter" data-level="1.2.2" data-path="index.html"><a href="index.html#r-level"><i class="fa fa-check"></i><b>1.2.2</b> <code>R</code> level</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#glossary"><i class="fa fa-check"></i><b>1.3</b> Glossary</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#how-does-this-work"><i class="fa fa-check"></i><b>1.4</b> How does this work?</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="index.html"><a href="index.html#use-your-own-data"><i class="fa fa-check"></i><b>1.4.1</b> Use your own data</a></li>
<li class="chapter" data-level="1.4.2" data-path="index.html"><a href="index.html#catching-up"><i class="fa fa-check"></i><b>1.4.2</b> Catching up</a></li>
<li class="chapter" data-level="1.4.3" data-path="index.html"><a href="index.html#tinker-time"><i class="fa fa-check"></i><b>1.4.3</b> Tinker time!</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#help-me"><i class="fa fa-check"></i><b>1.5</b> Help me!</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#references"><i class="fa fa-check"></i><b>1.6</b> References</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html"><i class="fa fa-check"></i><b>2</b> From observations to traitspaces</a>
<ul>
<li class="chapter" data-level="2.1" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#morphological-traits"><i class="fa fa-check"></i><b>2.1</b> Morphological traits</a>
<ul>
<li class="chapter" data-level="2.1.1" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#morphological-traits-in-the-age-of-molecular-data"><i class="fa fa-check"></i><b>2.1.1</b> Morphological traits in the age of molecular data?</a></li>
<li class="chapter" data-level="2.1.2" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#continuous-trait-data"><i class="fa fa-check"></i><b>2.1.2</b> Continuous trait data</a></li>
<li class="chapter" data-level="2.1.3" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#discrete-trait-data"><i class="fa fa-check"></i><b>2.1.3</b> Discrete trait data</a></li>
</ul></li>
<li class="chapter" data-level="2.2" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#ordinations"><i class="fa fa-check"></i><b>2.2</b> Ordinations</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#principal-components-analyses-pca"><i class="fa fa-check"></i><b>2.2.1</b> Principal components analyses (PCA)</a></li>
<li class="chapter" data-level="2.2.2" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#principal-coordinates-analysis-pco"><i class="fa fa-check"></i><b>2.2.2</b> Principal coordinates analysis (PCO)</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#estimating-ancestral-traits"><i class="fa fa-check"></i><b>2.3</b> Estimating ancestral traits</a>
<ul>
<li class="chapter" data-level="2.3.1" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#reading-trees-into-r"><i class="fa fa-check"></i><b>2.3.1</b> Reading trees into <code>R</code></a></li>
<li class="chapter" data-level="2.3.2" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#estimating-ancestral-traits-1"><i class="fa fa-check"></i><b>2.3.2</b> Estimating ancestral traits</a></li>
</ul></li>
<li class="chapter" data-level="2.4" data-path="from-observations-to-traitspaces.html"><a href="from-observations-to-traitspaces.html#references-1"><i class="fa fa-check"></i><b>2.4</b> References</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="from-traitspaces-to-disparity.html"><a href="from-traitspaces-to-disparity.html"><i class="fa fa-check"></i><b>3</b> From traitspaces to disparity</a>
<ul>
<li class="chapter" data-level="3.1" data-path="from-traitspaces-to-disparity.html"><a href="from-traitspaces-to-disparity.html#measuring-disparity"><i class="fa fa-check"></i><b>3.1</b> Measuring disparity</a></li>
<li class="chapter" data-level="3.2" data-path="from-traitspaces-to-disparity.html"><a href="from-traitspaces-to-disparity.html#the-curse-of-dimensionality"><i class="fa fa-check"></i><b>3.2</b> The curse of dimensionality</a>
<ul>
<li class="chapter" data-level="3.2.1" data-path="from-traitspaces-to-disparity.html"><a href="from-traitspaces-to-disparity.html#reducing-the-number-of-dimensions"><i class="fa fa-check"></i><b>3.2.1</b> Reducing the number of dimensions</a></li>
</ul></li>
<li class="chapter" data-level="3.3" data-path="from-traitspaces-to-disparity.html"><a href="from-traitspaces-to-disparity.html#which-disparity-metric-to-use"><i class="fa fa-check"></i><b>3.3</b> Which disparity metric to use?</a>
<ul>
<li class="chapter" data-level="3.3.1" data-path="from-traitspaces-to-disparity.html"><a href="from-traitspaces-to-disparity.html#choosing-and-testing-disparity-metrics"><i class="fa fa-check"></i><b>3.3.1</b> Choosing and testing disparity metrics</a></li>
</ul></li>
<li class="chapter" data-level="3.4" data-path="from-traitspaces-to-disparity.html"><a href="from-traitspaces-to-disparity.html#references-2"><i class="fa fa-check"></i><b>3.4</b> References</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html"><i class="fa fa-check"></i><b>4</b> From disparity to macroevolution</a>
<ul>
<li class="chapter" data-level="4.1" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#testing-hypothesis-with-disparity-data"><i class="fa fa-check"></i><b>4.1</b> Testing hypothesis with disparity data</a>
<ul>
<li class="chapter" data-level="4.1.1" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#making-subsets"><i class="fa fa-check"></i><b>4.1.1</b> Making subsets</a></li>
<li class="chapter" data-level="4.1.2" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#bootstrapping-and-rarefying"><i class="fa fa-check"></i><b>4.1.2</b> Bootstrapping and rarefying</a></li>
<li class="chapter" data-level="4.1.3" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#testing-hypotheses-1"><i class="fa fa-check"></i><b>4.1.3</b> Testing hypotheses 1</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#measuring-disparity-through-time"><i class="fa fa-check"></i><b>4.2</b> Measuring disparity through time</a>
<ul>
<li class="chapter" data-level="4.2.1" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#time-binning"><i class="fa fa-check"></i><b>4.2.1</b> Time binning</a></li>
<li class="chapter" data-level="4.2.2" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#time-slicing"><i class="fa fa-check"></i><b>4.2.2</b> Time slicing!</a></li>
<li class="chapter" data-level="4.2.3" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#dtt-disparity-through-time"><i class="fa fa-check"></i><b>4.2.3</b> <code>dtt</code>: disparity through time</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#measuring-disparity-with-trees"><i class="fa fa-check"></i><b>4.3</b> Measuring disparity with trees</a></li>
<li class="chapter" data-level="4.4" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#telling-disparity-stories"><i class="fa fa-check"></i><b>4.4</b> Telling disparity stories</a></li>
<li class="chapter" data-level="4.5" data-path="from-disparity-to-macroevolution.html"><a href="from-disparity-to-macroevolution.html#references-3"><i class="fa fa-check"></i><b>4.5</b> References</a></li>
</ul></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Traits data in R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="from-disparity-to-macroevolution" class="section level1 hasAnchor" number="4">
<h1><span class="header-section-number">4</span> From disparity to macroevolution<a href="from-disparity-to-macroevolution.html#from-disparity-to-macroevolution" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div class="warning" style="padding:0.1em; background-color:#FFAF70; color:#954200">
<p><strong>CATCHING UP ZONE</strong>: here’s some example discrete dataset from <span class="citation">Beck and Lee (<a href="#ref-beck2014ancient" role="doc-biblioref">2014</a>)</span> with ancestral states estimations and</p>
<details>
<p><summary>[<strong>Click to expand the data loading part</strong>]:
</summary></p>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb61-1"><a href="from-disparity-to-macroevolution.html#cb61-1" aria-hidden="true"></a><span class="co">## Get discrete data</span></span>
<span id="cb61-2"><a href="from-disparity-to-macroevolution.html#cb61-2" aria-hidden="true"></a>discrete_traits &lt;-<span class="st"> </span><span class="kw">read.nexus.data</span>(<span class="st">&quot;../examples/discrete_characters/Beck2014.nex&quot;</span>)</span>
<span id="cb61-3"><a href="from-disparity-to-macroevolution.html#cb61-3" aria-hidden="true"></a><span class="co">## Combine the output list into a matrix (rows are species and columns are characters)</span></span>
<span id="cb61-4"><a href="from-disparity-to-macroevolution.html#cb61-4" aria-hidden="true"></a>discrete_data &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, discrete_traits)</span>
<span id="cb61-5"><a href="from-disparity-to-macroevolution.html#cb61-5" aria-hidden="true"></a><span class="co">## Get the tree</span></span>
<span id="cb61-6"><a href="from-disparity-to-macroevolution.html#cb61-6" aria-hidden="true"></a>discrete_tree &lt;-<span class="st"> </span><span class="kw">read.nexus</span>(<span class="st">&quot;../examples/trees/Beck2014.tre&quot;</span>)</span>
<span id="cb61-7"><a href="from-disparity-to-macroevolution.html#cb61-7" aria-hidden="true"></a></span>
<span id="cb61-8"><a href="from-disparity-to-macroevolution.html#cb61-8" aria-hidden="true"></a><span class="co">## Cleaning both the tree and the data to match</span></span>
<span id="cb61-9"><a href="from-disparity-to-macroevolution.html#cb61-9" aria-hidden="true"></a>cleaned_data  &lt;-<span class="st"> </span><span class="kw">clean.data</span>(<span class="dt">data =</span> discrete_data, <span class="dt">tree =</span> discrete_tree)</span>
<span id="cb61-10"><a href="from-disparity-to-macroevolution.html#cb61-10" aria-hidden="true"></a>discrete_data &lt;-<span class="st"> </span>cleaned_data<span class="op">$</span>data</span>
<span id="cb61-11"><a href="from-disparity-to-macroevolution.html#cb61-11" aria-hidden="true"></a>discrete_tree &lt;-<span class="st"> </span>cleaned_data<span class="op">$</span>tree</span>
<span id="cb61-12"><a href="from-disparity-to-macroevolution.html#cb61-12" aria-hidden="true"></a><span class="co">## The following were dropped:</span></span>
<span id="cb61-13"><a href="from-disparity-to-macroevolution.html#cb61-13" aria-hidden="true"></a>cleaned_data<span class="op">$</span>dropped_rows</span></code></pre></div>
<pre><code>## [1] &quot;Montanalestes&quot;    &quot;Lainodon&quot;         &quot;Kharmerungulatum&quot; &quot;Alymlestes&quot;</code></pre>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb63-1"><a href="from-disparity-to-macroevolution.html#cb63-1" aria-hidden="true"></a>cleaned_data<span class="op">$</span>dropped_tips</span></code></pre></div>
<pre><code>## [1] NA</code></pre>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb65-1"><a href="from-disparity-to-macroevolution.html#cb65-1" aria-hidden="true"></a><span class="co">## Adding node labels</span></span>
<span id="cb65-2"><a href="from-disparity-to-macroevolution.html#cb65-2" aria-hidden="true"></a>discrete_tree &lt;-<span class="st"> </span><span class="kw">makeNodeLabel</span>(discrete_tree, <span class="dt">prefix =</span> <span class="st">&quot;n&quot;</span>)</span>
<span id="cb65-3"><a href="from-disparity-to-macroevolution.html#cb65-3" aria-hidden="true"></a><span class="co">## Adding the root time</span></span>
<span id="cb65-4"><a href="from-disparity-to-macroevolution.html#cb65-4" aria-hidden="true"></a>discrete_tree<span class="op">$</span>root.time &lt;-<span class="st"> </span><span class="kw">max</span>(<span class="kw">tree.age</span>(discrete_tree)<span class="op">$</span>age)</span>
<span id="cb65-5"><a href="from-disparity-to-macroevolution.html#cb65-5" aria-hidden="true"></a></span>
<span id="cb65-6"><a href="from-disparity-to-macroevolution.html#cb65-6" aria-hidden="true"></a><span class="co">## Estimating ancestral states</span></span>
<span id="cb65-7"><a href="from-disparity-to-macroevolution.html#cb65-7" aria-hidden="true"></a>ancestral_states &lt;-<span class="st"> </span><span class="kw">multi.ace</span>(discrete_data, discrete_tree, <span class="dt">output =</span> <span class="st">&quot;combined&quot;</span>)</span>
<span id="cb65-8"><a href="from-disparity-to-macroevolution.html#cb65-8" aria-hidden="true"></a></span>
<span id="cb65-9"><a href="from-disparity-to-macroevolution.html#cb65-9" aria-hidden="true"></a><span class="co"># Calculating the distance matrix</span></span>
<span id="cb65-10"><a href="from-disparity-to-macroevolution.html#cb65-10" aria-hidden="true"></a>distance_matrix &lt;-<span class="st"> </span><span class="kw">char.diff</span>(ancestral_states, <span class="dt">method =</span> <span class="st">&quot;mord&quot;</span>, <span class="dt">by.col =</span> <span class="ot">FALSE</span>)</span>
<span id="cb65-11"><a href="from-disparity-to-macroevolution.html#cb65-11" aria-hidden="true"></a></span>
<span id="cb65-12"><a href="from-disparity-to-macroevolution.html#cb65-12" aria-hidden="true"></a><span class="co">## Ordinating the data</span></span>
<span id="cb65-13"><a href="from-disparity-to-macroevolution.html#cb65-13" aria-hidden="true"></a>my_pco &lt;-<span class="st"> </span><span class="kw">cmdscale</span>(distance_matrix,</span>
<span id="cb65-14"><a href="from-disparity-to-macroevolution.html#cb65-14" aria-hidden="true"></a>                   <span class="co"># the number of dimensions</span></span>
<span id="cb65-15"><a href="from-disparity-to-macroevolution.html#cb65-15" aria-hidden="true"></a>                   <span class="dt">k =</span> <span class="kw">ncol</span>(distance_matrix)<span class="op">-</span><span class="dv">2</span>,</span>
<span id="cb65-16"><a href="from-disparity-to-macroevolution.html#cb65-16" aria-hidden="true"></a>                   <span class="co"># the Cailliez correction</span></span>
<span id="cb65-17"><a href="from-disparity-to-macroevolution.html#cb65-17" aria-hidden="true"></a>                   <span class="dt">add =</span> <span class="ot">TRUE</span>)</span>
<span id="cb65-18"><a href="from-disparity-to-macroevolution.html#cb65-18" aria-hidden="true"></a></span>
<span id="cb65-19"><a href="from-disparity-to-macroevolution.html#cb65-19" aria-hidden="true"></a><span class="co">## Getting just the coordinates</span></span>
<span id="cb65-20"><a href="from-disparity-to-macroevolution.html#cb65-20" aria-hidden="true"></a>my_traitspace &lt;-<span class="st"> </span>my_pco<span class="op">$</span>points</span></code></pre></div>
</details>
</div>
<p>Once we have our question at hand (and an idea of how to measure disparity) we can finally start doing some cool stuff and use disparity to tell macroevolutionary (or ecological!) stories.</p>
<div id="testing-hypothesis-with-disparity-data" class="section level2 hasAnchor" number="4.1">
<h2><span class="header-section-number">4.1</span> Testing hypothesis with disparity data<a href="from-disparity-to-macroevolution.html#testing-hypothesis-with-disparity-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One first aspect for these type of analyses, is to split the data in a meaningful way that will illustrate the process in question.
We can split it through time (we’ll cover that very soon) or by groups using the <code>dispRity::custom.subsets</code> function.</p>
<div id="making-subsets" class="section level3 hasAnchor" number="4.1.1">
<h3><span class="header-section-number">4.1.1</span> Making subsets<a href="from-disparity-to-macroevolution.html#making-subsets" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>This function is relatively easy to use, you can just provide your trait space and a specific way to group you data.
For example that could be two different clades or some lists of species with specific characteristics (e.g. diet).</p>
<div class="warning" style="padding:0.1em; background-color:#5D7BBB; color:#092663">
<p><strong>USE YOUR DATA</strong>: set up your groups based on your own data.</p>
<p>Either load a data frame (.csv) with different categories, for example:</p>
<pre><code>  , diet, group
species_a, omnivore, group A
species_b, omnivore, group B
species_c, carnivore, group A</code></pre>
<div class="sourceCode" id="cb67"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb67-1"><a href="from-disparity-to-macroevolution.html#cb67-1" aria-hidden="true"></a><span class="co">## Read your favourite csv file (you&#39;ll need to set up the path yourself!)</span></span>
<span id="cb67-2"><a href="from-disparity-to-macroevolution.html#cb67-2" aria-hidden="true"></a>my_data_groups &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;path_to_my_super_serious_dataset.csv&quot;</span>)</span></code></pre></div>
<p>or else you can just create your list manually</p>
<div class="sourceCode" id="cb68"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb68-1"><a href="from-disparity-to-macroevolution.html#cb68-1" aria-hidden="true"></a><span class="co">## Making a list of different groups</span></span>
<span id="cb68-2"><a href="from-disparity-to-macroevolution.html#cb68-2" aria-hidden="true"></a>my_data_groups &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;omnivores&quot;</span>  =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;species_a&quot;</span>, <span class="st">&quot;species_b&quot;</span>, ...),</span>
<span id="cb68-3"><a href="from-disparity-to-macroevolution.html#cb68-3" aria-hidden="true"></a>                       <span class="st">&quot;carnivores&quot;</span> =<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;species_c&quot;</span>, ...))</span></code></pre></div>
</div>
<p>Here we will just group the species from the example dataset into crown and stem mammals using the <code>dispRity::crown.stem</code> function:</p>
<div class="warning" style="padding:0.1em; background-color:#FFAF70; color:#954200">
<p><strong>CATCHING UP ZONE</strong>: Creating two groups from the <span class="citation">Beck and Lee (<a href="#ref-beck2014ancient" role="doc-biblioref">2014</a>)</span> dataset: the crown and the stem mammals</p>
<div class="sourceCode" id="cb69"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb69-1"><a href="from-disparity-to-macroevolution.html#cb69-1" aria-hidden="true"></a><span class="co">## Crown and stem mammals</span></span>
<span id="cb69-2"><a href="from-disparity-to-macroevolution.html#cb69-2" aria-hidden="true"></a>my_data_groups &lt;-<span class="st"> </span><span class="kw">crown.stem</span>(<span class="dt">tree =</span> discrete_tree, <span class="dt">inc.nodes =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
</div>
<p>And here’s how we can create the groups into a <code>"dispRity"</code> object:</p>
<div class="sourceCode" id="cb70"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb70-1"><a href="from-disparity-to-macroevolution.html#cb70-1" aria-hidden="true"></a><span class="co">## Grouping the data into different subsets</span></span>
<span id="cb70-2"><a href="from-disparity-to-macroevolution.html#cb70-2" aria-hidden="true"></a>my_groups &lt;-<span class="st"> </span><span class="kw">custom.subsets</span>(my_traitspace, <span class="dt">group =</span> my_data_groups)</span></code></pre></div>
<p>Note from here on we will be heavily relying on <code>"dispRity"</code> objects.
They are a specific <code>R</code> class of objects (S3 - for those that know what that is <a href="https://raw.githack.com/TGuillerme/dispRity/master/inst/vignettes/Developer_resources.html">more nerding here</a>) that are attached to the <code>dispRity</code> package.
It allows for easier visualisation (plotting, printing, summarising) and neater easier to reproduce analysis: each step is configured by you and is sharable and savable.</p>
<div class="sourceCode" id="cb71"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb71-1"><a href="from-disparity-to-macroevolution.html#cb71-1" aria-hidden="true"></a><span class="co">## What is the class of my_groups?</span></span>
<span id="cb71-2"><a href="from-disparity-to-macroevolution.html#cb71-2" aria-hidden="true"></a><span class="kw">class</span>(my_groups)</span></code></pre></div>
<pre><code>## [1] &quot;dispRity&quot;</code></pre>
<div class="sourceCode" id="cb73"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb73-1"><a href="from-disparity-to-macroevolution.html#cb73-1" aria-hidden="true"></a><span class="co">## What&#39;s in it?</span></span>
<span id="cb73-2"><a href="from-disparity-to-macroevolution.html#cb73-2" aria-hidden="true"></a>my_groups</span></code></pre></div>
<pre><code>##  ---- dispRity object ---- 
## 2 customised subsets for 203 elements in one matrix:
##     crown, stem.</code></pre>
<div class="sourceCode" id="cb75"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb75-1"><a href="from-disparity-to-macroevolution.html#cb75-1" aria-hidden="true"></a><span class="co">## What does it look like in 2D</span></span>
<span id="cb75-2"><a href="from-disparity-to-macroevolution.html#cb75-2" aria-hidden="true"></a><span class="kw">plot</span>(my_groups)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-45-1.png" width="672" /></p>
<p>Once we have this groups sorted, we can easily measure disparity:</p>
<div class="sourceCode" id="cb76"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb76-1"><a href="from-disparity-to-macroevolution.html#cb76-1" aria-hidden="true"></a><span class="co">## The sum of variance of our groups</span></span>
<span id="cb76-2"><a href="from-disparity-to-macroevolution.html#cb76-2" aria-hidden="true"></a>my_disparity &lt;-<span class="st"> </span><span class="kw">dispRity</span>(my_groups, <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb76-3"><a href="from-disparity-to-macroevolution.html#cb76-3" aria-hidden="true"></a>my_disparity</span></code></pre></div>
<pre><code>##  ---- dispRity object ---- 
## 2 customised subsets for 203 elements in one matrix with 201 dimensions:
##     crown, stem.
## Disparity was calculated as: c(sum, variances).</code></pre>
<div class="sourceCode" id="cb78"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb78-1"><a href="from-disparity-to-macroevolution.html#cb78-1" aria-hidden="true"></a><span class="co">## What&#39;s the difference in variance?</span></span>
<span id="cb78-2"><a href="from-disparity-to-macroevolution.html#cb78-2" aria-hidden="true"></a><span class="kw">summary</span>(my_disparity)</span></code></pre></div>
<pre><code>##   subsets   n   obs
## 1   crown 103 16.48
## 2    stem  99 16.36</code></pre>
<p>We can compare the values of disparity we have between our two groups and tell a story about it.
For example, with the example dataset we have a slight increase in variance in the crown mammals compared to the stem ones indicating maybe a increase in trait space occupancy through the phylogeny?
But that’s very anecdotal and needs more convincing.
We probably want to do some statistical test!</p>
<blockquote>
<p>A personal note on statistical testing for statophobes: we are now entering the zone when there’s going to be p-values and other statistical bits that can be a bit scary.
But fear not: although statistics (and researchers) often present this as an objective measure of the truth that has to be rigorous and final, I see it much more as a methodical way to convince our colleagues that our hypothesis is justifiable.
I think that statistics are not the truth but a very rigorous and shareable way to explain what is happening in the natural world.
A great paper to this effect is <span class="citation">Dushoff, Kain, and Bolker (<a href="#ref-dushoff2019can" role="doc-biblioref">2019</a>)</span> that neatly suggests changing the word “significant” (i.e. a magical black box word that makes your research correct or not) into the more commonly english word “clearly” (i.e. a normal word that does call any magic: “disparity in group A is clearly/not clearly different than in group B”).
Hopefully this approach will make you more relaxed about the following session and just see it as a <em>rigorously subjective</em> way to convince your comrades rather than some <em>objective</em> dark magic!
Note that note all my comrades (e.g. <span class="citation">Muff et al. (<a href="#ref-muff2022rewriting" role="doc-biblioref">2022</a>)</span>) share this view of statistics though ;).</p>
</blockquote>
</div>
<div id="bootstrapping-and-rarefying" class="section level3 hasAnchor" number="4.1.2">
<h3><span class="header-section-number">4.1.2</span> Bootstrapping and rarefying<a href="from-disparity-to-macroevolution.html#bootstrapping-and-rarefying" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>One way to check whether these two samples are truly different and not only different by chance is to bootstrap the data.
Bootstrapping is an easy way to look at the quality of your data (<span class="citation">Efron and Tibshirani (<a href="#ref-efron1994introduction" role="doc-biblioref">1994</a>)</span> [^3]).
Basically it consists of randomly resampling your data (with replacement) a certain number of times.
Effectively this means randomly increasing the weight of some of your data points (the ones resampled) and reducing the weight of some other (the ones not resampled).
[^3] If you’re ever struggling with stats and R, they made an excellent YouTube playlist of <a href="https://www.youtube.com/@datascienceanalytics1826/playlists">68 short stats courses</a>. Do one per day after lunch and you’ll be a stats expert in 3.5 months!</p>
<div class="sourceCode" id="cb80"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb80-1"><a href="from-disparity-to-macroevolution.html#cb80-1" aria-hidden="true"></a><span class="co">## Bootstrapping the data</span></span>
<span id="cb80-2"><a href="from-disparity-to-macroevolution.html#cb80-2" aria-hidden="true"></a>my_disparity &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(my_groups),</span>
<span id="cb80-3"><a href="from-disparity-to-macroevolution.html#cb80-3" aria-hidden="true"></a>                         <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb80-4"><a href="from-disparity-to-macroevolution.html#cb80-4" aria-hidden="true"></a>my_disparity</span></code></pre></div>
<pre><code>##  ---- dispRity object ---- 
## 2 customised subsets for 203 elements in one matrix with 201 dimensions:
##     crown, stem.
## Data was bootstrapped 100 times (method:&quot;full&quot;).
## Disparity was calculated as: c(sum, variances).</code></pre>
<div class="sourceCode" id="cb82"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb82-1"><a href="from-disparity-to-macroevolution.html#cb82-1" aria-hidden="true"></a><span class="co">## What the difference in variance after bootstrapping?</span></span>
<span id="cb82-2"><a href="from-disparity-to-macroevolution.html#cb82-2" aria-hidden="true"></a><span class="kw">summary</span>(my_disparity)</span></code></pre></div>
<pre><code>##   subsets   n   obs bs.median  2.5%   25%   75% 97.5%
## 1   crown 103 16.48     16.35 16.12 16.28 16.40 16.53
## 2    stem  99 16.36     16.22 16.06 16.15 16.26 16.34</code></pre>
<p>With the example dataset we see again a very small difference but with small-ish overlap in the confidence intervals:</p>
<div class="sourceCode" id="cb84"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb84-1"><a href="from-disparity-to-macroevolution.html#cb84-1" aria-hidden="true"></a><span class="co">## Visualising the difference in distribution</span></span>
<span id="cb84-2"><a href="from-disparity-to-macroevolution.html#cb84-2" aria-hidden="true"></a><span class="kw">plot</span>(my_disparity)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-48-1.png" width="672" /></p>
<p>This difference though might be due to the difference in number of species (more species in the crown group means more disparity?).
You can check that by <em>rarefying</em> your data.
This is a variant of the bootstrapping procedure but instead of resampling the same number of elements, you can set a specific number of elements to resample (or several numbers):</p>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb85-1"><a href="from-disparity-to-macroevolution.html#cb85-1" aria-hidden="true"></a><span class="co">## Get the size of the smallest group@</span></span>
<span id="cb85-2"><a href="from-disparity-to-macroevolution.html#cb85-2" aria-hidden="true"></a>n_small &lt;-<span class="st"> </span><span class="kw">min</span>(<span class="kw">size.subsets</span>(my_groups))</span>
<span id="cb85-3"><a href="from-disparity-to-macroevolution.html#cb85-3" aria-hidden="true"></a></span>
<span id="cb85-4"><a href="from-disparity-to-macroevolution.html#cb85-4" aria-hidden="true"></a><span class="co">## Bootstrapping the data</span></span>
<span id="cb85-5"><a href="from-disparity-to-macroevolution.html#cb85-5" aria-hidden="true"></a>my_disparity &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(my_groups, <span class="dt">rarefaction =</span> n_small),</span>
<span id="cb85-6"><a href="from-disparity-to-macroevolution.html#cb85-6" aria-hidden="true"></a>                         <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb85-7"><a href="from-disparity-to-macroevolution.html#cb85-7" aria-hidden="true"></a>my_disparity</span></code></pre></div>
<pre><code>##  ---- dispRity object ---- 
## 2 customised subsets for 203 elements in one matrix with 201 dimensions:
##     crown, stem.
## Data was bootstrapped 100 times (method:&quot;full&quot;) and rarefied to 99 elements.
## Disparity was calculated as: c(sum, variances).</code></pre>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb87-1"><a href="from-disparity-to-macroevolution.html#cb87-1" aria-hidden="true"></a><span class="co">## What the difference in variance after bootstrapping?</span></span>
<span id="cb87-2"><a href="from-disparity-to-macroevolution.html#cb87-2" aria-hidden="true"></a><span class="kw">summary</span>(my_disparity)</span></code></pre></div>
<pre><code>##   subsets   n   obs bs.median  2.5%   25%   75% 97.5%
## 1   crown 103 16.48     16.32 16.16 16.24 16.38 16.51
## 2   crown  99    NA     16.33 16.15 16.26 16.38 16.48
## 3    stem  99 16.36     16.22 16.06 16.15 16.26 16.34</code></pre>
<div class="sourceCode" id="cb89"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb89-1"><a href="from-disparity-to-macroevolution.html#cb89-1" aria-hidden="true"></a><span class="co">## Visualising the difference in distribution</span></span>
<span id="cb89-2"><a href="from-disparity-to-macroevolution.html#cb89-2" aria-hidden="true"></a><span class="kw">plot</span>(my_disparity)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-49-1.png" width="672" /></p>
<p>Here we can see it doesn’t make a big difference.</p>
<div class="warning" style="padding:0.1em; background-color:#65D85F; color:#067800">
<p><strong>TINKER TIMES</strong>: here we just looked at the <code>"full"</code> bootstrapping algorithm (that resamples every elements or every number of elements with the <code>rarefaction</code> option).
You can also change it to the more conservative <code>"single"</code> bootstrapping that resamples only one element every time; or the <code>"null"</code> algorithm that resamples <em>ALL</em> the elements in the trait space (i.e. not only the elements in the group).</p>
<p>Think about what the difference each algorithm could make to your data and test it!</p>
<details>
<p><summary>[<strong>Click to expand the solution</strong>]:
</summary></p>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb90-1"><a href="from-disparity-to-macroevolution.html#cb90-1" aria-hidden="true"></a><span class="co">## Bootstrapping the data 3 ways</span></span>
<span id="cb90-2"><a href="from-disparity-to-macroevolution.html#cb90-2" aria-hidden="true"></a>full_boot   &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(my_groups, <span class="dt">boot.type =</span> <span class="st">&quot;full&quot;</span>),</span>
<span id="cb90-3"><a href="from-disparity-to-macroevolution.html#cb90-3" aria-hidden="true"></a>                        <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb90-4"><a href="from-disparity-to-macroevolution.html#cb90-4" aria-hidden="true"></a>single_boot &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(my_groups, <span class="dt">boot.type =</span> <span class="st">&quot;single&quot;</span>),</span>
<span id="cb90-5"><a href="from-disparity-to-macroevolution.html#cb90-5" aria-hidden="true"></a>                        <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb90-6"><a href="from-disparity-to-macroevolution.html#cb90-6" aria-hidden="true"></a>null_boot   &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(my_groups, <span class="dt">boot.type =</span> <span class="st">&quot;null&quot;</span>),</span>
<span id="cb90-7"><a href="from-disparity-to-macroevolution.html#cb90-7" aria-hidden="true"></a>                        <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb90-8"><a href="from-disparity-to-macroevolution.html#cb90-8" aria-hidden="true"></a></span>
<span id="cb90-9"><a href="from-disparity-to-macroevolution.html#cb90-9" aria-hidden="true"></a><span class="co">## Plotting the differences</span></span>
<span id="cb90-10"><a href="from-disparity-to-macroevolution.html#cb90-10" aria-hidden="true"></a>op &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">3</span>))</span>
<span id="cb90-11"><a href="from-disparity-to-macroevolution.html#cb90-11" aria-hidden="true"></a><span class="kw">plot</span>(full_boot, <span class="dt">main =</span> <span class="st">&quot;full bootstrap&quot;</span>)</span>
<span id="cb90-12"><a href="from-disparity-to-macroevolution.html#cb90-12" aria-hidden="true"></a><span class="kw">plot</span>(single_boot, <span class="dt">main =</span> <span class="st">&quot;single bootstrap&quot;</span>)</span>
<span id="cb90-13"><a href="from-disparity-to-macroevolution.html#cb90-13" aria-hidden="true"></a><span class="kw">plot</span>(null_boot, <span class="dt">main =</span> <span class="st">&quot;null bootstrap&quot;</span>)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-50-1.png" width="1728" /></p>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb91-1"><a href="from-disparity-to-macroevolution.html#cb91-1" aria-hidden="true"></a><span class="kw">par</span>(op)</span></code></pre></div>
</details>
</div>
</div>
<div id="testing-hypotheses-1" class="section level3 hasAnchor" number="4.1.3">
<h3><span class="header-section-number">4.1.3</span> Testing hypotheses 1<a href="from-disparity-to-macroevolution.html#testing-hypotheses-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>Once we have a distribution of disparity values (here from our bootstrap), we can actually calculate the probability of our groups sharing the same distribution or not (i.e. calculating a <em>p</em>-value).
Because our distributions are not independent data (i.e. the values come from bootstraps, not observations), we can use a non-parametric t-test that doesn’t assume independence in the data: the Wilcoxon rank test (<code>wilcox.test</code>).
We can do that directly by extracting the bootstrapped disparity values (using get disparity):</p>
<div class="sourceCode" id="cb92"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb92-1"><a href="from-disparity-to-macroevolution.html#cb92-1" aria-hidden="true"></a><span class="co">## Extracting the disparity values</span></span>
<span id="cb92-2"><a href="from-disparity-to-macroevolution.html#cb92-2" aria-hidden="true"></a>bootstrapped_values &lt;-<span class="st"> </span><span class="kw">get.disparity</span>(my_disparity, <span class="dt">observed =</span> <span class="ot">FALSE</span>)</span>
<span id="cb92-3"><a href="from-disparity-to-macroevolution.html#cb92-3" aria-hidden="true"></a><span class="co">## Testing the differences</span></span>
<span id="cb92-4"><a href="from-disparity-to-macroevolution.html#cb92-4" aria-hidden="true"></a><span class="kw">wilcox.test</span>(bootstrapped_values<span class="op">$</span>crown[[<span class="dv">1</span>]], bootstrapped_values<span class="op">$</span>stem[[<span class="dv">1</span>]])</span></code></pre></div>
<pre><code>## 
##  Wilcoxon rank sum test with continuity correction
## 
## data:  bootstrapped_values$crown[[1]] and bootstrapped_values$stem[[1]]
## W = 8073, p-value = 6.035e-14
## alternative hypothesis: true location shift is not equal to 0</code></pre>
<p>Or, more nicely, by using the <code>test.dispRity</code> function:</p>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb94-1"><a href="from-disparity-to-macroevolution.html#cb94-1" aria-hidden="true"></a><span class="co">## Running the test automatically on the dispRity object</span></span>
<span id="cb94-2"><a href="from-disparity-to-macroevolution.html#cb94-2" aria-hidden="true"></a><span class="kw">test.dispRity</span>(my_disparity, <span class="dt">test =</span> wilcox.test)</span></code></pre></div>
<pre><code>## [[1]]
##              statistic: W
## crown : stem         8073
## 
## [[2]]
##                  p.value
## crown : stem 6.03536e-14</code></pre>
<blockquote>
<p>Note that you can also use disparity statistics that are distributions rather than point estimates. E.g. all the <code>variances</code> distribution instead of the <code>c(sum, variances)</code> point estimate. More info and examples <a href="https://raw.githack.com/TGuillerme/dispRity/master/inst/gitbook/_book/details-of-specific-functions.html#disparity-distribution">here</a>.</p>
</blockquote>
<p>If you had multiple groups, you can try doing pairs of tests using <code>comparisons</code> (but think about correcting for p-value inflation using the <code>correction</code> option), or, more neatly by using an <code>aov</code>.</p>
<blockquote>
<p>Note also that if you’re just interested in comparing the matrices for each subset (not their disparity), you can use the <code>adonis.dispRity</code> which is a PERMANOVA wrapper function for <code>vegan::adonis2</code>.</p>
</blockquote>
<!-- ### Testing hypotheses 3

#TODO!

@diaz2016global

Diaz style null testing
 -->
</div>
</div>
<div id="measuring-disparity-through-time" class="section level2 hasAnchor" number="4.2">
<h2><span class="header-section-number">4.2</span> Measuring disparity through time<a href="from-disparity-to-macroevolution.html#measuring-disparity-through-time" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>One other obvious way to look at changes in disparity is to look at disparity through time.
There are two main ways to look at it - which weirdly are both called “disparity through time”:</p>
<ul>
<li>Measuring disparity at different points in time. We are going to cover that first and call it simply <strong>disparity through time</strong>.</li>
<li>Measuring disparity through time and comparing it to models of change in disparity. We are going to cover that later and call it <strong><code>dtt</code></strong>.</li>
</ul>
<p>And we’ll chat about the differences between both methods of course!</p>
<div class="warning" style="padding:0.1em; background-color:#5D7BBB; color:#092663">
<p><strong>USE YOUR DATA</strong>: get the first and last occurrence data (FAD/LAD) for your own data (as a .csv file)</p>
<pre><code>  ,FAD ,LAD
species_a, 37.20, 36.80
species_b, 83.60, 72.10</code></pre>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb97-1"><a href="from-disparity-to-macroevolution.html#cb97-1" aria-hidden="true"></a><span class="co">## Read your favourite csv file (you&#39;ll need to set up the path yourself!)</span></span>
<span id="cb97-2"><a href="from-disparity-to-macroevolution.html#cb97-2" aria-hidden="true"></a>my_FADLAD &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;path_to_my_super_serious_dataset.csv&quot;</span>)</span></code></pre></div>
</div>
<div class="warning" style="padding:0.1em; background-color:#FFAF70; color:#954200">
<p><strong>CATCHING UP ZONE</strong>: Using the occurrence dataset from <span class="citation">Beck and Lee (<a href="#ref-beck2014ancient" role="doc-biblioref">2014</a>)</span></p>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb98-1"><a href="from-disparity-to-macroevolution.html#cb98-1" aria-hidden="true"></a><span class="co">## Crown and stem mammals</span></span>
<span id="cb98-2"><a href="from-disparity-to-macroevolution.html#cb98-2" aria-hidden="true"></a>my_FADLAD &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;../examples/FADLAD/Beck2014.csv&quot;</span>, <span class="dt">row.names =</span> <span class="dv">1</span>)</span></code></pre></div>
</div>
<div id="time-binning" class="section level3 hasAnchor" number="4.2.1">
<h3><span class="header-section-number">4.2.1</span> Time binning<a href="from-disparity-to-macroevolution.html#time-binning" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p>One first and relatively obvious way is to calculate disparity between groups but where groups are different time bins.
This is easily done in <code>dispRity</code> with the <code>chrono.subsets</code> function using the <code>method = "discrete"</code>.
Here for example, we can create 5 equally sized time bins across the temporal distribution of our data:</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb99-1"><a href="from-disparity-to-macroevolution.html#cb99-1" aria-hidden="true"></a><span class="co">## Creating discrete bin subsets</span></span>
<span id="cb99-2"><a href="from-disparity-to-macroevolution.html#cb99-2" aria-hidden="true"></a>time_subsets &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">method =</span> <span class="st">&quot;discrete&quot;</span>,</span>
<span id="cb99-3"><a href="from-disparity-to-macroevolution.html#cb99-3" aria-hidden="true"></a>                               <span class="dt">tree =</span> discrete_tree, <span class="dt">FADLAD =</span> my_FADLAD,</span>
<span id="cb99-4"><a href="from-disparity-to-macroevolution.html#cb99-4" aria-hidden="true"></a>                               <span class="dt">time =</span> <span class="dv">5</span>)</span></code></pre></div>
<blockquote>
<p>Note that if you have occurrence data for all your species you don’t need a <code>tree</code> argument.</p>
</blockquote>
<p>We can then calculate and visualise disparity</p>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb100-1"><a href="from-disparity-to-macroevolution.html#cb100-1" aria-hidden="true"></a><span class="co">## Calculating disparity per bin</span></span>
<span id="cb100-2"><a href="from-disparity-to-macroevolution.html#cb100-2" aria-hidden="true"></a>time_disparity &lt;-<span class="st"> </span><span class="kw">dispRity</span>(time_subsets, <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb100-3"><a href="from-disparity-to-macroevolution.html#cb100-3" aria-hidden="true"></a><span class="kw">summary</span>(time_disparity)</span></code></pre></div>
<pre><code>##           subsets  n   obs
## 1 168.35 - 134.68  3 16.52
## 2 134.68 - 101.01  7 16.64
## 3  101.01 - 67.34 31 16.72
## 4   67.34 - 33.67 50 17.26
## 5       33.67 - 0 18 17.33</code></pre>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb102-1"><a href="from-disparity-to-macroevolution.html#cb102-1" aria-hidden="true"></a><span class="kw">plot</span>(time_disparity)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-56-1.png" width="672" /></p>
<p>Tada! That was super easy!</p>
<div class="warning" style="padding:0.1em; background-color:#65D85F; color:#067800">
<p><strong>TINKER TIMES</strong>: you can also make it a little bit more interesting or at least a bit more related to geology by specifying time bins that correspond to geological layers.
You can do that by specifying a vector for time in the format of <code>time = c(145, 66, 23)</code> for the boundaries of the Cretaceous (from 145 million years - mya - to 66 mya) and the Palaeogene (66 to 23 mya). See if you can do that for your dataset.</p>
<details>
<p><summary>[<strong>Click to expand the solution</strong>]:
</summary></p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb103-1"><a href="from-disparity-to-macroevolution.html#cb103-1" aria-hidden="true"></a><span class="co">## Option 1: enter the boundaries ages manually</span></span>
<span id="cb103-2"><a href="from-disparity-to-macroevolution.html#cb103-2" aria-hidden="true"></a><span class="co">## booooooring! - and prone to errors</span></span>
<span id="cb103-3"><a href="from-disparity-to-macroevolution.html#cb103-3" aria-hidden="true"></a></span>
<span id="cb103-4"><a href="from-disparity-to-macroevolution.html#cb103-4" aria-hidden="true"></a><span class="co">## Option 2: use the palaeoverse package!</span></span>
<span id="cb103-5"><a href="from-disparity-to-macroevolution.html#cb103-5" aria-hidden="true"></a><span class="co">## Getting the geological time from the palaeoverse package</span></span>
<span id="cb103-6"><a href="from-disparity-to-macroevolution.html#cb103-6" aria-hidden="true"></a>period_bins &lt;-<span class="st"> </span>palaeoverse<span class="op">::</span><span class="kw">time_bins</span>(<span class="dt">rank =</span> <span class="st">&quot;period&quot;</span>)</span></code></pre></div>
<pre><code>## Registered S3 method overwritten by &#39;e1071&#39;:
##   method       from
##   print.fclust GET</code></pre>
<div class="sourceCode" id="cb105"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb105-1"><a href="from-disparity-to-macroevolution.html#cb105-1" aria-hidden="true"></a><span class="co">## Get the boundaries from Jurassic to present</span></span>
<span id="cb105-2"><a href="from-disparity-to-macroevolution.html#cb105-2" aria-hidden="true"></a>period_bins_bounds &lt;-<span class="st"> </span><span class="kw">c</span>(period_bins[<span class="kw">c</span>(<span class="dv">8</span><span class="op">:</span><span class="dv">12</span>), <span class="st">&quot;max_ma&quot;</span>], <span class="dv">0</span>)</span>
<span id="cb105-3"><a href="from-disparity-to-macroevolution.html#cb105-3" aria-hidden="true"></a></span>
<span id="cb105-4"><a href="from-disparity-to-macroevolution.html#cb105-4" aria-hidden="true"></a><span class="co">## Creating the subsets</span></span>
<span id="cb105-5"><a href="from-disparity-to-macroevolution.html#cb105-5" aria-hidden="true"></a>time_subsets_geo &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">method =</span> <span class="st">&quot;discrete&quot;</span>,</span>
<span id="cb105-6"><a href="from-disparity-to-macroevolution.html#cb105-6" aria-hidden="true"></a>                                  <span class="dt">tree =</span> discrete_tree, <span class="dt">FADLAD =</span> my_FADLAD,</span>
<span id="cb105-7"><a href="from-disparity-to-macroevolution.html#cb105-7" aria-hidden="true"></a>                                  <span class="dt">time =</span> period_bins_bounds)</span></code></pre></div>
<pre><code>## Warning: The interval 23.03 - 2.58 is empty.</code></pre>
<div class="sourceCode" id="cb107"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb107-1"><a href="from-disparity-to-macroevolution.html#cb107-1" aria-hidden="true"></a><span class="co">## Calculating disparity per bin</span></span>
<span id="cb107-2"><a href="from-disparity-to-macroevolution.html#cb107-2" aria-hidden="true"></a>time_disparity2 &lt;-<span class="st"> </span><span class="kw">dispRity</span>(time_subsets_geo, <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span></code></pre></div>
<pre><code>## Warning in dispRity(time_subsets_geo, metric = c(sum, variances)): Disparity
## not calculated for subset 23.03 - 2.58 (not enough data).</code></pre>
<div class="sourceCode" id="cb109"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb109-1"><a href="from-disparity-to-macroevolution.html#cb109-1" aria-hidden="true"></a><span class="kw">summary</span>(time_disparity2)</span></code></pre></div>
<pre><code>##        subsets  n   obs
## 1  201.3 - 145  3 16.52
## 2     145 - 66 40 16.86
## 3   66 - 23.03 50 17.31
## 4 23.03 - 2.58  0    NA
## 5     2.58 - 0 14 17.24</code></pre>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb111-1"><a href="from-disparity-to-macroevolution.html#cb111-1" aria-hidden="true"></a><span class="kw">plot</span>(time_disparity2)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-57-1.png" width="1728" /></p>
</details>
</div>
<p>This is all well and good but there can be some problems with the time binning method covered just above.
We discuss most of the potential problems here: <span class="citation">Guillerme and Cooper (<a href="#ref-guillerme2018time" role="doc-biblioref">2018</a>)</span>.
Two of the main ones are that:</p>
<ol style="list-style-type: decimal">
<li>It assumes punctuated equilibrium disparity evolution (i.e. disparity is uniform within bins and changes only - and drastically - at bins boundary)</li>
<li>It can artificially increase the effect of extinction events by forcing changes at a bin boundary.</li>
</ol>
</div>
<div id="time-slicing" class="section level3 hasAnchor" number="4.2.2">
<h3><span class="header-section-number">4.2.2</span> Time slicing!<a href="from-disparity-to-macroevolution.html#time-slicing" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><a href="https://raw.githack.com/TGuillerme/Morphological_traits/main/slides/03.1_time_slicing.pdf">SLIDES: Time slicing</a></p>
<p>Let’s try to do a simple ACCTRAN model (accelerated transition) on ten time slices:</p>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb112-1"><a href="from-disparity-to-macroevolution.html#cb112-1" aria-hidden="true"></a><span class="co">## Creating continuous time slices subsets</span></span>
<span id="cb112-2"><a href="from-disparity-to-macroevolution.html#cb112-2" aria-hidden="true"></a>time_subsets &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">method =</span> <span class="st">&quot;continuous&quot;</span>,</span>
<span id="cb112-3"><a href="from-disparity-to-macroevolution.html#cb112-3" aria-hidden="true"></a>                               <span class="dt">tree =</span> discrete_tree, <span class="dt">FADLAD =</span> my_FADLAD,</span>
<span id="cb112-4"><a href="from-disparity-to-macroevolution.html#cb112-4" aria-hidden="true"></a>                               <span class="dt">time =</span> <span class="dv">10</span>, <span class="dt">model =</span> <span class="st">&quot;acctran&quot;</span>)</span>
<span id="cb112-5"><a href="from-disparity-to-macroevolution.html#cb112-5" aria-hidden="true"></a><span class="co">## Calculating disparity per bin</span></span>
<span id="cb112-6"><a href="from-disparity-to-macroevolution.html#cb112-6" aria-hidden="true"></a>time_disparity &lt;-<span class="st"> </span><span class="kw">dispRity</span>(time_subsets, <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb112-7"><a href="from-disparity-to-macroevolution.html#cb112-7" aria-hidden="true"></a><span class="kw">summary</span>(time_disparity)</span></code></pre></div>
<pre><code>##    subsets  n   obs
## 1   168.35  3 15.79
## 2   149.64  6 16.22
## 3   130.94 14 16.38
## 4   112.23 18 16.22
## 5    93.53 26 16.44
## 6    74.82 36 16.41
## 7    56.12 38 17.09
## 8    37.41 24 17.41
## 9    18.71 15 17.28
## 10       0 14 17.24</code></pre>
<div class="sourceCode" id="cb114"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb114-1"><a href="from-disparity-to-macroevolution.html#cb114-1" aria-hidden="true"></a><span class="kw">plot</span>(time_disparity)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-58-1.png" width="672" /></p>
<p>Note that so far the model we use is not probabilistic, that means that every time you run it you’ll have the same solution.
However, some of the models (called the <code>*.split</code> models) are probabilistic in nature and would require you bootstrapping the data to show the range of estimated solutions:</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb115-1"><a href="from-disparity-to-macroevolution.html#cb115-1" aria-hidden="true"></a><span class="co">## Creating continuous time slices subsets using a split model</span></span>
<span id="cb115-2"><a href="from-disparity-to-macroevolution.html#cb115-2" aria-hidden="true"></a>gradual_split &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">method =</span> <span class="st">&quot;continuous&quot;</span>,</span>
<span id="cb115-3"><a href="from-disparity-to-macroevolution.html#cb115-3" aria-hidden="true"></a>                               <span class="dt">tree =</span> discrete_tree, <span class="dt">FADLAD =</span> my_FADLAD,</span>
<span id="cb115-4"><a href="from-disparity-to-macroevolution.html#cb115-4" aria-hidden="true"></a>                               <span class="dt">time =</span> <span class="dv">10</span>, <span class="dt">model =</span> <span class="st">&quot;gradual.split&quot;</span>)</span>
<span id="cb115-5"><a href="from-disparity-to-macroevolution.html#cb115-5" aria-hidden="true"></a><span class="co">## Calculating disparity per bin</span></span>
<span id="cb115-6"><a href="from-disparity-to-macroevolution.html#cb115-6" aria-hidden="true"></a>time_disparity &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(time_subsets), <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb115-7"><a href="from-disparity-to-macroevolution.html#cb115-7" aria-hidden="true"></a><span class="kw">summary</span>(time_disparity)</span></code></pre></div>
<pre><code>##    subsets  n   obs bs.median  2.5%   25%   75% 97.5%
## 1   168.35  3 15.79     10.81  0.00 10.15 15.79 15.79
## 2   149.64  6 16.22     13.25  8.84 12.09 14.10 15.38
## 3   130.94 14 16.38     15.31 14.09 15.08 15.56 15.94
## 4   112.23 18 16.22     15.37 14.55 15.19 15.56 15.86
## 5    93.53 26 16.44     15.85 15.31 15.72 15.99 16.11
## 6    74.82 36 16.41     15.97 15.63 15.84 16.08 16.21
## 7    56.12 38 17.09     16.67 16.28 16.56 16.75 16.88
## 8    37.41 24 17.41     16.71 16.15 16.53 16.84 16.99
## 9    18.71 15 17.28     16.20 15.22 15.94 16.45 16.76
## 10       0 14 17.24     16.11 15.04 15.85 16.33 16.83</code></pre>
<div class="sourceCode" id="cb117"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb117-1"><a href="from-disparity-to-macroevolution.html#cb117-1" aria-hidden="true"></a><span class="kw">plot</span>(time_disparity)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-59-1.png" width="672" /></p>
<div class="warning" style="padding:0.1em; background-color:#65D85F; color:#067800">
<p><strong>TINKER TIMES</strong>: have a look at comparing the different models of time-slicing and see how it changes your disparity curves (if it changes them at all!).</p>
<details>
<p><summary>[<strong>Click to expand the solution</strong>]:
</summary></p>
<div class="sourceCode" id="cb118"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb118-1"><a href="from-disparity-to-macroevolution.html#cb118-1" aria-hidden="true"></a><span class="co">## Testing different models</span></span>
<span id="cb118-2"><a href="from-disparity-to-macroevolution.html#cb118-2" aria-hidden="true"></a>gradual_split &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">method =</span> <span class="st">&quot;continuous&quot;</span>,</span>
<span id="cb118-3"><a href="from-disparity-to-macroevolution.html#cb118-3" aria-hidden="true"></a>                               <span class="dt">tree =</span> discrete_tree, <span class="dt">FADLAD =</span> my_FADLAD,</span>
<span id="cb118-4"><a href="from-disparity-to-macroevolution.html#cb118-4" aria-hidden="true"></a>                               <span class="dt">time =</span> <span class="dv">10</span>, <span class="dt">model =</span> <span class="st">&quot;gradual.split&quot;</span>)</span>
<span id="cb118-5"><a href="from-disparity-to-macroevolution.html#cb118-5" aria-hidden="true"></a>equal_split   &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">method =</span> <span class="st">&quot;continuous&quot;</span>,</span>
<span id="cb118-6"><a href="from-disparity-to-macroevolution.html#cb118-6" aria-hidden="true"></a>                               <span class="dt">tree =</span> discrete_tree, <span class="dt">FADLAD =</span> my_FADLAD,</span>
<span id="cb118-7"><a href="from-disparity-to-macroevolution.html#cb118-7" aria-hidden="true"></a>                               <span class="dt">time =</span> <span class="dv">10</span>, <span class="dt">model =</span> <span class="st">&quot;equal.split&quot;</span>)</span>
<span id="cb118-8"><a href="from-disparity-to-macroevolution.html#cb118-8" aria-hidden="true"></a>acctran       &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">method =</span> <span class="st">&quot;continuous&quot;</span>,</span>
<span id="cb118-9"><a href="from-disparity-to-macroevolution.html#cb118-9" aria-hidden="true"></a>                               <span class="dt">tree =</span> discrete_tree, <span class="dt">FADLAD =</span> my_FADLAD,</span>
<span id="cb118-10"><a href="from-disparity-to-macroevolution.html#cb118-10" aria-hidden="true"></a>                               <span class="dt">time =</span> <span class="dv">10</span>, <span class="dt">model =</span> <span class="st">&quot;acctran&quot;</span>)</span>
<span id="cb118-11"><a href="from-disparity-to-macroevolution.html#cb118-11" aria-hidden="true"></a>deltran       &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">method =</span> <span class="st">&quot;continuous&quot;</span>,</span>
<span id="cb118-12"><a href="from-disparity-to-macroevolution.html#cb118-12" aria-hidden="true"></a>                               <span class="dt">tree =</span> discrete_tree, <span class="dt">FADLAD =</span> my_FADLAD,</span>
<span id="cb118-13"><a href="from-disparity-to-macroevolution.html#cb118-13" aria-hidden="true"></a>                               <span class="dt">time =</span> <span class="dv">10</span>, <span class="dt">model =</span> <span class="st">&quot;deltran&quot;</span>)</span>
<span id="cb118-14"><a href="from-disparity-to-macroevolution.html#cb118-14" aria-hidden="true"></a><span class="co">## Calculating disparity per bin</span></span>
<span id="cb118-15"><a href="from-disparity-to-macroevolution.html#cb118-15" aria-hidden="true"></a>gradual_split_disp &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(gradual_split), <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb118-16"><a href="from-disparity-to-macroevolution.html#cb118-16" aria-hidden="true"></a>equal_split_disp   &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(equal_split), <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb118-17"><a href="from-disparity-to-macroevolution.html#cb118-17" aria-hidden="true"></a>acctran_disp       &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(acctran), <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span>
<span id="cb118-18"><a href="from-disparity-to-macroevolution.html#cb118-18" aria-hidden="true"></a>deltran_disp       &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(deltran), <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances))</span></code></pre></div>
<pre><code>## Warning in boot.matrix(deltran): The following subsets have less than 3 elements: 168.35.
## This might effect the bootstrap/rarefaction output.</code></pre>
<div class="sourceCode" id="cb120"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb120-1"><a href="from-disparity-to-macroevolution.html#cb120-1" aria-hidden="true"></a>op &lt;-<span class="st"> </span><span class="kw">par</span>(<span class="dt">mfrow =</span> <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">2</span>))</span>
<span id="cb120-2"><a href="from-disparity-to-macroevolution.html#cb120-2" aria-hidden="true"></a><span class="kw">plot</span>(gradual_split_disp, <span class="dt">main =</span> <span class="st">&quot;gradual split&quot;</span>)</span>
<span id="cb120-3"><a href="from-disparity-to-macroevolution.html#cb120-3" aria-hidden="true"></a><span class="kw">plot</span>(equal_split_disp, <span class="dt">main =</span> <span class="st">&quot;equal split&quot;</span>)</span>
<span id="cb120-4"><a href="from-disparity-to-macroevolution.html#cb120-4" aria-hidden="true"></a><span class="kw">plot</span>(acctran_disp, <span class="dt">main =</span> <span class="st">&quot;acctran&quot;</span>)</span>
<span id="cb120-5"><a href="from-disparity-to-macroevolution.html#cb120-5" aria-hidden="true"></a><span class="kw">plot</span>(deltran_disp, <span class="dt">main =</span> <span class="st">&quot;deltran&quot;</span>)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-60-1.png" width="1152" /></p>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb121-1"><a href="from-disparity-to-macroevolution.html#cb121-1" aria-hidden="true"></a><span class="kw">par</span>(op)</span></code></pre></div>
</details>
</div>
</div>
<div id="dtt-disparity-through-time" class="section level3 hasAnchor" number="4.2.3">
<h3><span class="header-section-number">4.2.3</span> <code>dtt</code>: disparity through time<a href="from-disparity-to-macroevolution.html#dtt-disparity-through-time" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p><code>dtt</code> analyses are disparity through time analyses where the disparity is calculated for each clade in a given tree and then compared to a simulated disparity curve (usually a Brownian motion one) using the same tree.
This results in a comparison between an observed clade disparity curve scaled by the number of species and a simulated trait.
Note that this is quiet different than disparity through time how we measured it previously.</p>
<p>Here is a comparison with the three methods with:</p>
<ul>
<li>DTT being the disparity-through-time as described in <span class="citation">Harmon et al. (<a href="#ref-harmon2003tempo" role="doc-biblioref">2003</a>)</span> and calculated in <span class="citation">Pennell et al. (<a href="#ref-geiger" role="doc-biblioref">2014</a>)</span>;</li>
<li>the raw disparity per clade calculated using <code>dispRity</code> (and plotted through time).</li>
<li>the disparity through time using <code>dispRity</code> (time-slicing using the proximity model).</li>
</ul>
<blockquote>
<p>Note that I’m not going to go in the details here but you can see and follow the code in the raw markdown file if you want.</p>
</blockquote>
<pre><code>## Warning in dispRity(time_subsets, metric = average.sq): Disparity not
## calculated for subset 0.5833 (not enough data).</code></pre>
<p><img src="trait_data_files/figure-html/unnamed-chunk-61-1.png" width="672" /></p>
<p>You can see there is some difference between all the curves here!
The ones between the DTT and the clade disparity (black and blue) can be probably explained by algorithmic choices between both approaches (the scaling procedure is a bit more complex in <code>geiger:dtt</code> and the clade disparity values are time slice averages - from what I understood).
But in general both curves are similar.</p>
<p>This is not the case though with the disparity through time analyses <code>dispRity</code> style where disparity is measured as snapshots of disparity in the trait space at any specific time point (here the nodes ages, corresponding to the clade ages).
So keep these two differences in mind and sorry about the confusing name similarities!</p>
<p>So which method is better?
I have some opinion here but I’m biased and I like to remind you to choose the method best adapted to your question, not especially because of somebody’s opinion (e.g. mine).</p>
<p>That said, one of the major drawbacks I see with the <code>dtt</code> method is that, unfortunately, the implementation is not really modular.
It currently doesn’t allow trees with fossil data and it only works with three inbuilt disparity metrics.</p>
<p>In general though, both methods (<code>dtt</code> and disparity through time) both have the problem on ways to test null expectations that are easily to calculate but also make sense biologically.
For example, in <code>dtt</code>, comparing the observed disparity to a simple Brownian Motion can often be a bit naive and unsurprising (i.e. it’s often easily expected that some trait have a complex evolutionary history and are not just following a BM).</p>
<p>Stay tuned for some potential solution with the <span class="citation">Guillerme (<a href="#ref-treats" role="doc-biblioref">2024</a>)</span> package allowing to simulate more complex data and evolutionary histories to compare observed and simulated disparity.
I’m still working on it but if you feel explorative, please contact me or check out this loooooong tutorial <a href="https://www.youtube.com/watch?v=QUhaFZxER2I">video</a> and <a href="https://cdn.githubraw.com/TGuillerme/treats/master/inst/vignettes/treats_workshop.html">vignette</a>.</p>
</div>
</div>
<div id="measuring-disparity-with-trees" class="section level2 hasAnchor" number="4.3">
<h2><span class="header-section-number">4.3</span> Measuring disparity with trees<a href="from-disparity-to-macroevolution.html#measuring-disparity-with-trees" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>Finally, note that here we haven’t looked much at phylogenetic hypotheses and phylogenetic comparative methods.
This is mainly because I didn’t want to add to much stuff and that we’ll have a great session dedicated to it tomorrow!</p>
<p>But here’s a brief example on how to look at disparity using standard phylogenetic generalised least square (pgls) method.
Note that this type of analysis is only relevant if you calculated disparity using a dimension level 2 metric that is linked to the rows (for example the distance of each element to the centroid using <code>centroids</code>) and if you have only data for your tips:</p>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb123-1"><a href="from-disparity-to-macroevolution.html#cb123-1" aria-hidden="true"></a><span class="co">## Simple example</span></span>
<span id="cb123-2"><a href="from-disparity-to-macroevolution.html#cb123-2" aria-hidden="true"></a><span class="kw">data</span>(BeckLee_mat50)</span>
<span id="cb123-3"><a href="from-disparity-to-macroevolution.html#cb123-3" aria-hidden="true"></a><span class="kw">data</span>(BeckLee_tree)</span>
<span id="cb123-4"><a href="from-disparity-to-macroevolution.html#cb123-4" aria-hidden="true"></a>disparity &lt;-<span class="st"> </span><span class="kw">dispRity</span>(BeckLee_mat50, <span class="dt">metric =</span> centroids, <span class="dt">tree =</span> BeckLee_tree)</span>
<span id="cb123-5"><a href="from-disparity-to-macroevolution.html#cb123-5" aria-hidden="true"></a></span>
<span id="cb123-6"><a href="from-disparity-to-macroevolution.html#cb123-6" aria-hidden="true"></a><span class="co">## Running a simple PGLS</span></span>
<span id="cb123-7"><a href="from-disparity-to-macroevolution.html#cb123-7" aria-hidden="true"></a>model &lt;-<span class="st"> </span><span class="kw">pgls.dispRity</span>(disparity)</span>
<span id="cb123-8"><a href="from-disparity-to-macroevolution.html#cb123-8" aria-hidden="true"></a><span class="kw">summary</span>(model)</span></code></pre></div>
<pre><code>## 
## Call:
## [1] &quot;dispRity interface of phylolm using: formula = disparity ~ 1 and model = BM&quot;
## 
##     AIC  logLik 
## -158.71   81.35 
## 
## Raw residuals:
##      Min       1Q   Median       3Q      Max 
## -0.06247 -0.02075  0.03188  0.07433  0.14999 
## 
## Mean tip height: 88.13181
## Parameter estimate(s) using ML:
## sigma2: 5.776737e-05 
## 
## Coefficients:
##             Estimate   StdErr t.value   p.value    
## (Intercept) 1.541096 0.024797  62.148 &lt; 2.2e-16 ***
## ---
## Signif. codes:  0 &#39;***&#39; 0.001 &#39;**&#39; 0.01 &#39;*&#39; 0.05 &#39;.&#39; 0.1 &#39; &#39; 1
## 
## R-squared:     0 Adjusted R-squared:     0</code></pre>
<p>We will be covering that in much more details tomorrow!</p>
</div>
<div id="telling-disparity-stories" class="section level2 hasAnchor" number="4.4">
<h2><span class="header-section-number">4.4</span> Telling disparity stories<a href="from-disparity-to-macroevolution.html#telling-disparity-stories" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>We’ve looked at many methods but now it’s time to work on telling a story!</p>
<p>Here are some examples:</p>
<ul>
<li><a href="https://www.biorxiv.org/content/10.1101/2024.06.27.600807v1.abstract">Morphological evolutionary bursts in skinks</a> by <span class="citation">Brennan et al. (<a href="#ref-brennan2024evolutionary" role="doc-biblioref">2024</a>)</span></li>
</ul>
<details>
<p><summary>[<strong>Click here to read in the data</strong>]:
</summary></p>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb125-1"><a href="from-disparity-to-macroevolution.html#cb125-1" aria-hidden="true"></a><span class="co">## Loading the data from Brennan 2024 (you will need to do the PCA yourself!)</span></span>
<span id="cb125-2"><a href="from-disparity-to-macroevolution.html#cb125-2" aria-hidden="true"></a>data &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;../examples/continuous_characters/Brennan2024.csv&quot;</span>) </span>
<span id="cb125-3"><a href="from-disparity-to-macroevolution.html#cb125-3" aria-hidden="true"></a></span>
<span id="cb125-4"><a href="from-disparity-to-macroevolution.html#cb125-4" aria-hidden="true"></a><span class="co">## And here&#39;s the tree</span></span>
<span id="cb125-5"><a href="from-disparity-to-macroevolution.html#cb125-5" aria-hidden="true"></a>tree &lt;-<span class="st"> </span><span class="kw">read.tree</span>(<span class="st">&quot;../examples/trees/Brennan2024.tre&quot;</span>) </span></code></pre></div>
</details>
<ul>
<li><a href="https://royalsocietypublishing.org/doi/full/10.1098/rspb.2014.1278">Mammals around the K-Pg extinction</a> by <span class="citation">Beck and Lee (<a href="#ref-beck2014ancient" role="doc-biblioref">2014</a>)</span></li>
</ul>
<details>
<p><summary>[<strong>Click here to read in the data</strong>]:
</summary></p>
<div class="sourceCode" id="cb126"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb126-1"><a href="from-disparity-to-macroevolution.html#cb126-1" aria-hidden="true"></a><span class="co">## Loading the data from Beck and Lee 2014 (you will need to do the PCO yourself!)</span></span>
<span id="cb126-2"><a href="from-disparity-to-macroevolution.html#cb126-2" aria-hidden="true"></a>data &lt;-<span class="st"> </span><span class="kw">read.nexus.data</span>(<span class="st">&quot;../examples/discrete_characters/Beck2014.nex&quot;</span>) </span>
<span id="cb126-3"><a href="from-disparity-to-macroevolution.html#cb126-3" aria-hidden="true"></a></span>
<span id="cb126-4"><a href="from-disparity-to-macroevolution.html#cb126-4" aria-hidden="true"></a><span class="co">## And here&#39;s the tree</span></span>
<span id="cb126-5"><a href="from-disparity-to-macroevolution.html#cb126-5" aria-hidden="true"></a>tree &lt;-<span class="st"> </span><span class="kw">read.nexus</span>(<span class="st">&quot;../examples/trees/Beck2014.tre&quot;</span>)</span>
<span id="cb126-6"><a href="from-disparity-to-macroevolution.html#cb126-6" aria-hidden="true"></a></span>
<span id="cb126-7"><a href="from-disparity-to-macroevolution.html#cb126-7" aria-hidden="true"></a><span class="co">## And the tip ages</span></span>
<span id="cb126-8"><a href="from-disparity-to-macroevolution.html#cb126-8" aria-hidden="true"></a>FADLAD &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;../examples/FADLAD/Beck2014.csv&quot;</span>, <span class="dt">row.names =</span> <span class="dv">1</span>) </span></code></pre></div>
</details>
<ul>
<li><a href="https://www.cambridge.org/core/journals/journal-of-paleontology/article/bayesian-estimation-of-fossil-phylogenies-and-the-evolution-of-early-to-middle-paleozoic-crinoids-echinodermata/E37972902541CD0995AAD08A1122BD54">Echinoderms during the Ordovician and Devonian</a> by <span class="citation">Wright (<a href="#ref-wright2017bayesian" role="doc-biblioref">2017</a>)</span></li>
</ul>
<details>
<p><summary>[<strong>Click here to read in the data</strong>]:
</summary></p>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb127-1"><a href="from-disparity-to-macroevolution.html#cb127-1" aria-hidden="true"></a><span class="co">## Loading the data from Wright 2017 (you will need to do the PCO yourself!)</span></span>
<span id="cb127-2"><a href="from-disparity-to-macroevolution.html#cb127-2" aria-hidden="true"></a>data &lt;-<span class="st"> </span><span class="kw">read.nexus.data</span>(<span class="st">&quot;../examples/discrete_characters/Wright2017.nex&quot;</span>) </span>
<span id="cb127-3"><a href="from-disparity-to-macroevolution.html#cb127-3" aria-hidden="true"></a></span>
<span id="cb127-4"><a href="from-disparity-to-macroevolution.html#cb127-4" aria-hidden="true"></a><span class="co">## And here&#39;s the tree</span></span>
<span id="cb127-5"><a href="from-disparity-to-macroevolution.html#cb127-5" aria-hidden="true"></a>tree &lt;-<span class="st"> </span><span class="kw">read.nexus</span>(<span class="st">&quot;../examples/trees/Wright2017.tre&quot;</span>)</span>
<span id="cb127-6"><a href="from-disparity-to-macroevolution.html#cb127-6" aria-hidden="true"></a></span>
<span id="cb127-7"><a href="from-disparity-to-macroevolution.html#cb127-7" aria-hidden="true"></a><span class="co">## And the tip ages</span></span>
<span id="cb127-8"><a href="from-disparity-to-macroevolution.html#cb127-8" aria-hidden="true"></a>FADLAD &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;../examples/FADLAD/Wright2017.csv&quot;</span>, <span class="dt">row.names =</span> <span class="dv">1</span>) </span></code></pre></div>
</details>
<ul>
<li><a href="https://www.sciencedirect.com/science/article/pii/S0960982214010471">Dinosaurs - bird transition</a> by <span class="citation">Brusatte et al. (<a href="#ref-brusatte2014gradual" role="doc-biblioref">2014</a>)</span></li>
</ul>
<details>
<p><summary>[<strong>Click here to read in the data</strong>]:
</summary></p>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb128-1"><a href="from-disparity-to-macroevolution.html#cb128-1" aria-hidden="true"></a><span class="co">## Loading the data from Beck and Lee 2014 (you will need to do the PCO yourself!)</span></span>
<span id="cb128-2"><a href="from-disparity-to-macroevolution.html#cb128-2" aria-hidden="true"></a>data &lt;-<span class="st"> </span><span class="kw">read.nexus.data</span>(<span class="st">&quot;../examples/discrete_characters/Brusatte2014.nex&quot;</span>) </span>
<span id="cb128-3"><a href="from-disparity-to-macroevolution.html#cb128-3" aria-hidden="true"></a></span>
<span id="cb128-4"><a href="from-disparity-to-macroevolution.html#cb128-4" aria-hidden="true"></a><span class="co">## And here&#39;s the tree</span></span>
<span id="cb128-5"><a href="from-disparity-to-macroevolution.html#cb128-5" aria-hidden="true"></a>tree &lt;-<span class="st"> </span><span class="kw">read.nexus</span>(<span class="st">&quot;../examples/trees/Brusatte2014.nex&quot;</span>)</span>
<span id="cb128-6"><a href="from-disparity-to-macroevolution.html#cb128-6" aria-hidden="true"></a></span>
<span id="cb128-7"><a href="from-disparity-to-macroevolution.html#cb128-7" aria-hidden="true"></a><span class="co">## And the tip ages</span></span>
<span id="cb128-8"><a href="from-disparity-to-macroevolution.html#cb128-8" aria-hidden="true"></a>FADLAD &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="st">&quot;../examples/FADLAD/Brusatte2014.csv&quot;</span>, <span class="dt">row.names =</span> <span class="dv">1</span>) </span></code></pre></div>
</details>
<ul>
<li><a href="https://onlinelibrary.wiley.com/doi/full/10.1111/jeb.13540">disparity through time in crocodiles</a> by <span class="citation">Godoy (<a href="#ref-godoy2020crocodylomorph" role="doc-biblioref">2020</a>)</span></li>
</ul>
<details>
<p><summary>[<strong>Click here to read in the data</strong>]:
</summary></p>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb129-1"><a href="from-disparity-to-macroevolution.html#cb129-1" aria-hidden="true"></a><span class="kw">load</span>(<span class="st">&quot;../examples/gmm_characters/Godoy2019.RData&quot;</span>) </span>
<span id="cb129-2"><a href="from-disparity-to-macroevolution.html#cb129-2" aria-hidden="true"></a><span class="co">## landmark data</span></span>
<span id="cb129-3"><a href="from-disparity-to-macroevolution.html#cb129-3" aria-hidden="true"></a><span class="co"># landmark_data</span></span>
<span id="cb129-4"><a href="from-disparity-to-macroevolution.html#cb129-4" aria-hidden="true"></a><span class="co"># semi_landmarks</span></span>
<span id="cb129-5"><a href="from-disparity-to-macroevolution.html#cb129-5" aria-hidden="true"></a></span>
<span id="cb129-6"><a href="from-disparity-to-macroevolution.html#cb129-6" aria-hidden="true"></a><span class="co">## Tree</span></span>
<span id="cb129-7"><a href="from-disparity-to-macroevolution.html#cb129-7" aria-hidden="true"></a>tree &lt;-<span class="st"> </span>alter1_MCCT</span>
<span id="cb129-8"><a href="from-disparity-to-macroevolution.html#cb129-8" aria-hidden="true"></a>GPA &lt;-<span class="st"> </span>geomorph<span class="op">::</span><span class="kw">gpagen</span>(landmark_data, <span class="dt">curves =</span> semi_landmarks, <span class="dt">surfaces =</span> <span class="ot">NULL</span>, <span class="dt">PrinAxes =</span> <span class="ot">TRUE</span>, <span class="dt">max.iter =</span> <span class="ot">NULL</span>, <span class="dt">ProcD =</span> F, <span class="dt">Proj =</span> <span class="ot">TRUE</span>, <span class="dt">print.progress =</span> <span class="ot">TRUE</span>)</span></code></pre></div>
<pre><code>## 
## Performing GPA
##   |                                                                              |                                                                      |   0%  |                                                                              |=======                                                               |  10%  |                                                                              |==============                                                        |  20%  |                                                                              |=====================                                                 |  30%  |                                                                              |============================                                          |  40%  |                                                                              |===================================                                   |  50%  |                                                                              |==========================================                            |  60%  |                                                                              |=================================================                     |  70%  |                                                                              |========================================================              |  80%  |                                                                              |===============================================================       |  90%  |                                                                              |======================================================================| 100%
## 
## Making projections... Finished!</code></pre>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb131-1"><a href="from-disparity-to-macroevolution.html#cb131-1" aria-hidden="true"></a>traitspace &lt;-<span class="st"> </span>dispRity<span class="op">::</span><span class="kw">geomorph.ordination</span>(GPA)</span></code></pre></div>
</details>
<ul>
<li><a href="https://onlinelibrary.wiley.com/doi/full/10.1002/ece3.10201">relation between changes in disparity and temperature in turtles</a> by <span class="citation">Farina et al. (<a href="#ref-farina2023turtle" role="doc-biblioref">2023</a>)</span></li>
</ul>
<details>
<p><summary>[<strong>Click here to read in the data</strong>]:
</summary></p>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb132-1"><a href="from-disparity-to-macroevolution.html#cb132-1" aria-hidden="true"></a><span class="co">## Loading the data from Farina 2023</span></span>
<span id="cb132-2"><a href="from-disparity-to-macroevolution.html#cb132-2" aria-hidden="true"></a><span class="kw">load</span>(<span class="st">&quot;../examples/continuous_characters/Farina2023.RData&quot;</span>) </span>
<span id="cb132-3"><a href="from-disparity-to-macroevolution.html#cb132-3" aria-hidden="true"></a></span>
<span id="cb132-4"><a href="from-disparity-to-macroevolution.html#cb132-4" aria-hidden="true"></a><span class="co">## This is the age of each tip</span></span>
<span id="cb132-5"><a href="from-disparity-to-macroevolution.html#cb132-5" aria-hidden="true"></a>FADLAD &lt;-<span class="st"> </span>fbd_ages</span>
<span id="cb132-6"><a href="from-disparity-to-macroevolution.html#cb132-6" aria-hidden="true"></a><span class="co">## This is a list of trees (note that you can do dispRity analyses with multiple trees!)</span></span>
<span id="cb132-7"><a href="from-disparity-to-macroevolution.html#cb132-7" aria-hidden="true"></a>tree_list &lt;-<span class="st"> </span>Sterli_10trees</span>
<span id="cb132-8"><a href="from-disparity-to-macroevolution.html#cb132-8" aria-hidden="true"></a><span class="co">## Data and factors (unidimensional)</span></span>
<span id="cb132-9"><a href="from-disparity-to-macroevolution.html#cb132-9" aria-hidden="true"></a>traitspace_table &lt;-<span class="st"> </span>data_turtles</span></code></pre></div>
</details>
<p>Other cool examples:</p>
<ul>
<li><a href="https://www.nature.com/articles/s41559-024-02491-9">morphospace changes in brachiopods</a> by <span class="citation">Guo et al. (<a href="#ref-guo2024morphological" role="doc-biblioref">2024</a>)</span></li>
<li><a href="https://www.cell.com/current-biology/fulltext/S0960-9822(24)00529-3#%20">diversification of sabertooth tigers</a> by <span class="citation">Chatar et al. (<a href="#ref-chatar2024evolutionary" role="doc-biblioref">2024</a>)</span></li>
<li><a href="https://royalsocietypublishing.org/doi/full/10.1098/rsos.231495">Mezosoic archosaurs disparity</a> by <span class="citation">Shipley et al. (<a href="#ref-shipley2024locomotion" role="doc-biblioref">2024</a>)</span></li>
<li><a href="https://www.cambridge.org/core/journals/paleobiology/article/convergence-and-constraint-in-the-cranial-evolution-of-mosasaurid-reptiles-and-early-cetaceans/19DC42DECB1C9166732122F1857EAC5F">convergence between mosasaurs and early cetacean skulls</a> by <span class="citation">Bennion et al. (<a href="#ref-bennion2023convergence" role="doc-biblioref">2023</a>)</span></li>
<li><a href="https://royalsocietypublishing.org/doi/full/10.1098/rsos.231951">niche partitioning in marine reptiles</a> by <span class="citation">Foffa, Young, and Brusatte (<a href="#ref-foffa2024comparative" role="doc-biblioref">2024</a>)</span></li>
</ul>
<div class="warning" style="padding:0.1em; background-color:#65D85F; color:#067800">
<p><strong>TINKER TIMES</strong>: Do your very own disparity analysis!</p>
<p>Think about what the difference each algorithm could make to your data and test it!</p>
<details>
<p><summary>[<strong>Click to expand the solution</strong>]:
</summary></p>
<p>Question: is there an effect of the the K-Pg extinction on mammalian disparity?</p>
<p>Here we will be looking at how the</p>
<p><strong>1- From data to trait spaces</strong></p>
<p>Here we will again use the data from <span class="citation">Beck and Lee (<a href="#ref-beck2014ancient" role="doc-biblioref">2014</a>)</span> to test our hypothesis.
We will transform the data by doing:</p>
<ul>
<li>Ancestral states estimations on discrete characters</li>
<li>Calculating a distance matrix using the MORD distance (<span class="citation">Lloyd (<a href="#ref-lloyd2016estimating" role="doc-biblioref">2016</a>)</span>)</li>
<li>Ordinating the distance matrix using a PCO with the Cailliez correction (<span class="citation">Cailliez (<a href="#ref-cailliez1983analytical" role="doc-biblioref">1983</a>)</span>)</li>
</ul>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb133-1"><a href="from-disparity-to-macroevolution.html#cb133-1" aria-hidden="true"></a><span class="co">## Get discrete data</span></span>
<span id="cb133-2"><a href="from-disparity-to-macroevolution.html#cb133-2" aria-hidden="true"></a>discrete_traits &lt;-<span class="st"> </span><span class="kw">read.nexus.data</span>(<span class="st">&quot;../examples/discrete_characters/Beck2014.nex&quot;</span>)</span>
<span id="cb133-3"><a href="from-disparity-to-macroevolution.html#cb133-3" aria-hidden="true"></a><span class="co">## Combine the output list into a matrix (rows are species and columns are characters)</span></span>
<span id="cb133-4"><a href="from-disparity-to-macroevolution.html#cb133-4" aria-hidden="true"></a>discrete_data &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, discrete_traits)</span>
<span id="cb133-5"><a href="from-disparity-to-macroevolution.html#cb133-5" aria-hidden="true"></a><span class="co">## Get the tree</span></span>
<span id="cb133-6"><a href="from-disparity-to-macroevolution.html#cb133-6" aria-hidden="true"></a>discrete_tree &lt;-<span class="st"> </span><span class="kw">read.nexus</span>(<span class="st">&quot;../examples/trees/Beck2014.tre&quot;</span>)</span>
<span id="cb133-7"><a href="from-disparity-to-macroevolution.html#cb133-7" aria-hidden="true"></a></span>
<span id="cb133-8"><a href="from-disparity-to-macroevolution.html#cb133-8" aria-hidden="true"></a><span class="co">## Cleaning both the tree and the data to match</span></span>
<span id="cb133-9"><a href="from-disparity-to-macroevolution.html#cb133-9" aria-hidden="true"></a>cleaned_data  &lt;-<span class="st"> </span><span class="kw">clean.data</span>(<span class="dt">data =</span> discrete_data, <span class="dt">tree =</span> discrete_tree)</span>
<span id="cb133-10"><a href="from-disparity-to-macroevolution.html#cb133-10" aria-hidden="true"></a>discrete_data &lt;-<span class="st"> </span>cleaned_data<span class="op">$</span>data</span>
<span id="cb133-11"><a href="from-disparity-to-macroevolution.html#cb133-11" aria-hidden="true"></a>discrete_tree &lt;-<span class="st"> </span>cleaned_data<span class="op">$</span>tree</span>
<span id="cb133-12"><a href="from-disparity-to-macroevolution.html#cb133-12" aria-hidden="true"></a></span>
<span id="cb133-13"><a href="from-disparity-to-macroevolution.html#cb133-13" aria-hidden="true"></a><span class="co">## Adding node labels</span></span>
<span id="cb133-14"><a href="from-disparity-to-macroevolution.html#cb133-14" aria-hidden="true"></a>discrete_tree &lt;-<span class="st"> </span><span class="kw">makeNodeLabel</span>(discrete_tree, <span class="dt">prefix =</span> <span class="st">&quot;n&quot;</span>)</span>
<span id="cb133-15"><a href="from-disparity-to-macroevolution.html#cb133-15" aria-hidden="true"></a><span class="co">## Adding the root time</span></span>
<span id="cb133-16"><a href="from-disparity-to-macroevolution.html#cb133-16" aria-hidden="true"></a>discrete_tree<span class="op">$</span>root.time &lt;-<span class="st"> </span><span class="kw">max</span>(<span class="kw">tree.age</span>(discrete_tree)<span class="op">$</span>age)</span>
<span id="cb133-17"><a href="from-disparity-to-macroevolution.html#cb133-17" aria-hidden="true"></a></span>
<span id="cb133-18"><a href="from-disparity-to-macroevolution.html#cb133-18" aria-hidden="true"></a><span class="co">## Estimating ancestral states</span></span>
<span id="cb133-19"><a href="from-disparity-to-macroevolution.html#cb133-19" aria-hidden="true"></a>ancestral_states &lt;-<span class="st"> </span><span class="kw">multi.ace</span>(discrete_data, discrete_tree, <span class="dt">output =</span> <span class="st">&quot;combined&quot;</span>)</span>
<span id="cb133-20"><a href="from-disparity-to-macroevolution.html#cb133-20" aria-hidden="true"></a></span>
<span id="cb133-21"><a href="from-disparity-to-macroevolution.html#cb133-21" aria-hidden="true"></a><span class="co"># Calculating the distance matrix</span></span>
<span id="cb133-22"><a href="from-disparity-to-macroevolution.html#cb133-22" aria-hidden="true"></a>distance_matrix &lt;-<span class="st"> </span><span class="kw">char.diff</span>(ancestral_states, <span class="dt">method =</span> <span class="st">&quot;mord&quot;</span>, <span class="dt">by.col =</span> <span class="ot">FALSE</span>)</span>
<span id="cb133-23"><a href="from-disparity-to-macroevolution.html#cb133-23" aria-hidden="true"></a></span>
<span id="cb133-24"><a href="from-disparity-to-macroevolution.html#cb133-24" aria-hidden="true"></a><span class="co">## Ordinating the data</span></span>
<span id="cb133-25"><a href="from-disparity-to-macroevolution.html#cb133-25" aria-hidden="true"></a>my_pco &lt;-<span class="st"> </span><span class="kw">cmdscale</span>(distance_matrix,</span>
<span id="cb133-26"><a href="from-disparity-to-macroevolution.html#cb133-26" aria-hidden="true"></a>                   <span class="co"># the number of dimensions</span></span>
<span id="cb133-27"><a href="from-disparity-to-macroevolution.html#cb133-27" aria-hidden="true"></a>                   <span class="dt">k =</span> <span class="kw">ncol</span>(distance_matrix)<span class="op">-</span><span class="dv">2</span>,</span>
<span id="cb133-28"><a href="from-disparity-to-macroevolution.html#cb133-28" aria-hidden="true"></a>                   <span class="co"># the Cailliez correction</span></span>
<span id="cb133-29"><a href="from-disparity-to-macroevolution.html#cb133-29" aria-hidden="true"></a>                   <span class="dt">add =</span> <span class="ot">TRUE</span>)</span>
<span id="cb133-30"><a href="from-disparity-to-macroevolution.html#cb133-30" aria-hidden="true"></a></span>
<span id="cb133-31"><a href="from-disparity-to-macroevolution.html#cb133-31" aria-hidden="true"></a><span class="co">## Getting just the coordinates</span></span>
<span id="cb133-32"><a href="from-disparity-to-macroevolution.html#cb133-32" aria-hidden="true"></a>my_traitspace &lt;-<span class="st"> </span>my_pco<span class="op">$</span>points</span></code></pre></div>
<p><strong>2- From traitspace to disparity</strong></p>
<p>From this traitspace we will choose one metric looking at the change in disparity trait space size using the sum of each dimension’s variance.</p>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb134-1"><a href="from-disparity-to-macroevolution.html#cb134-1" aria-hidden="true"></a><span class="co">## Testing the sum of variances as a proxy of traitspace size</span></span>
<span id="cb134-2"><a href="from-disparity-to-macroevolution.html#cb134-2" aria-hidden="true"></a>my_test &lt;-<span class="st"> </span><span class="kw">test.metric</span>(my_traitspace, <span class="dt">metric =</span> <span class="kw">c</span>(sum, variances), <span class="dt">shifts =</span> <span class="kw">c</span>(<span class="st">&quot;size&quot;</span>, <span class="st">&quot;random&quot;</span>))</span>
<span id="cb134-3"><a href="from-disparity-to-macroevolution.html#cb134-3" aria-hidden="true"></a><span class="kw">plot</span>(my_test)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-70-1.png" width="576" /></p>
<p><strong>3 - From disparity to macroevolution</strong></p>
<p>With the trait space and the chosen disparity metrics we can now test whether there is a change it trait space size or density before and after the K-Pg extinction?</p>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb135-1"><a href="from-disparity-to-macroevolution.html#cb135-1" aria-hidden="true"></a><span class="co">## Creating the time subsets</span></span>
<span id="cb135-2"><a href="from-disparity-to-macroevolution.html#cb135-2" aria-hidden="true"></a>time_subsets &lt;-<span class="st"> </span><span class="kw">chrono.subsets</span>(my_traitspace, <span class="dt">tree =</span> discrete_tree,</span>
<span id="cb135-3"><a href="from-disparity-to-macroevolution.html#cb135-3" aria-hidden="true"></a>                               <span class="dt">method =</span> <span class="st">&quot;continuous&quot;</span>, <span class="dt">model =</span> <span class="st">&quot;proximity&quot;</span>,</span>
<span id="cb135-4"><a href="from-disparity-to-macroevolution.html#cb135-4" aria-hidden="true"></a>                               <span class="dt">time =</span> <span class="kw">rev</span>(<span class="kw">seq</span>(<span class="dt">from =</span> <span class="dv">5</span>, <span class="dt">to =</span> <span class="dv">125</span>, <span class="dt">by =</span> <span class="dv">10</span>)))</span>
<span id="cb135-5"><a href="from-disparity-to-macroevolution.html#cb135-5" aria-hidden="true"></a><span class="co">## Visualising the traitspace</span></span>
<span id="cb135-6"><a href="from-disparity-to-macroevolution.html#cb135-6" aria-hidden="true"></a><span class="kw">plot</span>(time_subsets)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-71-1.png" width="576" /></p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb136-1"><a href="from-disparity-to-macroevolution.html#cb136-1" aria-hidden="true"></a><span class="co">## Measuring disparity through time</span></span>
<span id="cb136-2"><a href="from-disparity-to-macroevolution.html#cb136-2" aria-hidden="true"></a>disparity_size &lt;-<span class="st"> </span><span class="kw">dispRity</span>(<span class="kw">boot.matrix</span>(time_subsets), <span class="dt">metric =</span> <span class="kw">c</span>(median, centroids))</span>
<span id="cb136-3"><a href="from-disparity-to-macroevolution.html#cb136-3" aria-hidden="true"></a><span class="kw">plot</span>(disparity_size, <span class="dt">main =</span> <span class="st">&quot;Change in trait space size&quot;</span>)</span>
<span id="cb136-4"><a href="from-disparity-to-macroevolution.html#cb136-4" aria-hidden="true"></a><span class="kw">abline</span>(<span class="dt">v =</span> <span class="fl">7.2</span>, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>)</span></code></pre></div>
<p><img src="trait_data_files/figure-html/unnamed-chunk-72-1.png" width="1152" /></p>
<p>Testing the change in disparity in terms of of size after the K-Pg extinction with a lag of up to 30 million years:</p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb137-1"><a href="from-disparity-to-macroevolution.html#cb137-1" aria-hidden="true"></a><span class="co">## Selecting the two subsets after the mass extinction</span></span>
<span id="cb137-2"><a href="from-disparity-to-macroevolution.html#cb137-2" aria-hidden="true"></a>extinction_effect &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="kw">c</span>(<span class="st">&quot;65&quot;</span>, <span class="st">&quot;55&quot;</span>), <span class="kw">c</span>(<span class="st">&quot;65&quot;</span>, <span class="st">&quot;45&quot;</span>))</span>
<span id="cb137-3"><a href="from-disparity-to-macroevolution.html#cb137-3" aria-hidden="true"></a></span>
<span id="cb137-4"><a href="from-disparity-to-macroevolution.html#cb137-4" aria-hidden="true"></a><span class="co">## Testing the differences in disparity</span></span>
<span id="cb137-5"><a href="from-disparity-to-macroevolution.html#cb137-5" aria-hidden="true"></a>results &lt;-<span class="st"> </span><span class="kw">test.dispRity</span>(disparity_size, <span class="dt">comparisons =</span> extinction_effect,</span>
<span id="cb137-6"><a href="from-disparity-to-macroevolution.html#cb137-6" aria-hidden="true"></a>                         <span class="dt">test =</span> wilcox.test, <span class="dt">correction =</span> <span class="st">&quot;bonferroni&quot;</span>)</span>
<span id="cb137-7"><a href="from-disparity-to-macroevolution.html#cb137-7" aria-hidden="true"></a>results</span></code></pre></div>
<pre><code>## [[1]]
##         statistic: W
## 65 : 55         4997
## 65 : 45         9438
## 
## [[2]]
##              p.value
## 65 : 55 1.000000e+00
## 65 : 45 4.327658e-27</code></pre>
<p>Here we see now clear difference between before the K-Pg extinction (65 mya) and right after (55 mya) but some clear affect 20 million years later (45 mya).
This suggest more an effect of the PETM 55 mya than the K-Pg on this dataset!</p>
</details>
</div>
</div>
<div id="references-3" class="section level2 hasAnchor" number="4.5">
<h2><span class="header-section-number">4.5</span> References<a href="from-disparity-to-macroevolution.html#references-3" class="anchor-section" aria-label="Anchor link to header"></a></h2>

<div id="refs" class="references hanging-indent">
<div>
<p>Baken, E. K., M. L. Collyer, A. Kaliontzopoulou, and D. C. Adams. 2021. “Geomorph V4.0 and gmShiny: Enhanced Analytics and a New Graphical Interface for a Comprehensive Morphometric Experience.” <em>Methods in Ecology and Evolution</em> 12: 2355–63.</p>
</div>
<div>
<p>Bapst, David W. 2012. “Paleotree: An R Package for Paleontological and Phylogenetic Analyses of Evolution.” <em>Methods in Ecology and Evolution</em> 3 (5): 803–7.</p>
</div>
<div>
<p>Beck, Robin MD, and Michael SY Lee. 2014. “Ancient Dates or Accelerated Rates? Morphological Clocks and the Antiquity of Placental Mammals.” <em>Proceedings of the Royal Society B: Biological Sciences</em> 281 (1793): 20141278.</p>
</div>
<div>
<p>Bell, Mark A., and Graeme T. Lloyd. 2015. “Strap: An R Package for Plotting Phylogenies Against Stratigraphy and Assessing Their Stratigraphic Congruence.” <em>Palaeontology</em> 58: 379–89. <a href="https://doi.org/10.1111/pala.12142">https://doi.org/10.1111/pala.12142</a>.</p>
</div>
<div>
<p>Bennion, Rebecca F, Jamie A MacLaren, Ellen J Coombs, Felix G Marx, Olivier Lambert, and Valentin Fischer. 2023. “Convergence and Constraint in the Cranial Evolution of Mosasaurid Reptiles and Early Cetaceans.” <em>Paleobiology</em> 49 (2): 215–31.</p>
</div>
<div>
<p>Brazeau, Martin D. 2011. “Problematic Character Coding Methods in Morphology and Their Effects.” <em>Biological Journal of the Linnean Society</em> 104 (3): 489–98.</p>
</div>
<div>
<p>Brennan, Ian G, David G Chapple, J Scott Keogh, and Stephen Donnellan. 2024. “Evolutionary Bursts Drive Morphological Novelty in the World’s Largest Skinks.” <em>bioRxiv</em>, 2024–06.</p>
</div>
<div>
<p>Brusatte, Stephen L, Graeme T Lloyd, Steve C Wang, and Mark A Norell. 2014. “Gradual Assembly of Avian Body Plan Culminated in Rapid Rates of Evolution Across the Dinosaur-Bird Transition.” <em>Current Biology</em> 24 (20): 2386–92.</p>
</div>
<div>
<p>Cailliez, Francis. 1983. “The Analytical Solution of the Additive Constant Problem.” <em>Psychometrika</em> 48 (2): 305–8.</p>
</div>
<div>
<p>Chatar, Narimane, Margot Michaud, Davide Tamagnini, and Valentin Fischer. 2024. “Evolutionary Patterns of Cat-Like Carnivorans Unveil Drivers of the Sabertooth Morphology.” <em>Current Biology</em> 34 (11): 2460–73.</p>
</div>
<div>
<p>Ciampaglio, Charles N, Matthieu Kemp, and Daniel W McShea. 2001. “Detecting Changes in Morphospace Occupation Patterns in the Fossil Record: Characterization and Analysis of Measures of Disparity.” <em>Paleobiology</em> 27 (4): 695–715.</p>
</div>
<div>
<p>Dushoff, Jonathan, Morgan P Kain, and Benjamin M Bolker. 2019. “I Can See Clearly Now: Reinterpreting Statistical Significance.” <em>Methods in Ecology and Evolution</em> 10 (6): 756–59.</p>
</div>
<div>
<p>Efron, Bradley, and Robert J Tibshirani. 1994. <em>An Introduction to the Bootstrap</em>. Chapman; Hall/CRC.</p>
</div>
<div>
<p>Farina, Bruna M, Pedro L Godoy, Roger BJ Benson, Max C Langer, and Gabriel S Ferreira. 2023. “Turtle Body Size Evolution Is Determined by Lineage-Specific Specializations Rather Than Global Trends.” <em>Ecology and Evolution</em> 13 (6): e10201.</p>
</div>
<div>
<p>Foffa, Davide, Mark T Young, and Stephen L Brusatte. 2024. “Comparative Functional Morphology Indicates Niche Partitioning Among Sympatric Marine Reptiles.” <em>Royal Society Open Science</em> 11 (5): 231951.</p>
</div>
<div>
<p>Gearty, William. 2024. <em>Deeptime: Plotting Tools for Anyone Working in Deep Time</em>. <a href="https://CRAN.R-project.org/package=deeptime">https://CRAN.R-project.org/package=deeptime</a>.</p>
</div>
<div>
<p>Godoy, Pedro L. 2020. “Crocodylomorph Cranial Shape Evolution and Its Relationship with Body Size and Ecology.” <em>Journal of Evolutionary Biology</em> 33 (1): 4–21.</p>
</div>
<div>
<p>Guillerme, Thomas. 2018. “DispRity: A Modular R Package for Measuring Disparity.” <em>Methods in Ecology and Evolution</em> 9 (7): 1755–63.</p>
</div>
<div>
<p>———. 2024. “Treats: A Modular R Package for Simulating Trees and Traits.” <em>Methods in Ecology and Evolution</em> 15 (4): 647–56.</p>
</div>
<div>
<p>Guillerme, Thomas, Pedro Cardoso, Maria Wagner Jørgensen, Stefano Mammola, and Thomas J Matthews. 2024. “The What, How and Why of Trait-Based Analyses in Ecology.” <em>bioRxiv</em>, 2024–06.</p>
</div>
<div>
<p>Guillerme, Thomas, and Natalie Cooper. 2018. “Time for a Rethink: Time Sub-Sampling Methods in Disparity-Through-Time Analyses.” <em>Palaeontology</em> 61 (4): 481–93.</p>
</div>
<div>
<p>Guillerme, Thomas, Natalie Cooper, Stephen L Brusatte, Katie E Davis, Andrew L Jackson, Sylvain Gerber, Anjali Goswami, et al. 2020. “Disparities in the Analysis of Morphological Disparity.” <em>Biology Letters</em> 16 (7): 20200199.</p>
</div>
<div>
<p>Guillerme, Thomas, Mark N Puttick, Ariel E Marcy, and Vera Weisbecker. 2020. “Shifting Spaces: Which Disparity or Dissimilarity Measurement Best Summarize Occupancy in Multidimensional Spaces?” <em>Ecology and Evolution</em> 10 (14): 7261–75.</p>
</div>
<div>
<p>Guo, Zhen, Michael J Benton, Thomas L Stubbs, and Zhong-Qiang Chen. 2024. “Morphological Innovation Did Not Drive Diversification in Mesozoic–Cenozoic Brachiopods.” <em>Nature Ecology &amp; Evolution</em>, 1–11.</p>
</div>
<div>
<p>Harmon, Luke J, James A Schulte, Allan Larson, and Jonathan B Losos. 2003. “Tempo and Mode of Evolutionary Radiation in Iguanian Lizards.” <em>Science</em> 301 (5635): 961–64.</p>
</div>
<div>
<p>Hopkins, Melanie J, and Sylvain Gerber. 2021. “Morphological Disparity.” <em>Evolutionary Developmental Biology: A Reference Guide</em>, 965–76.</p>
</div>
<div>
<p>Jones, Lewis A., William Gearty, Bethany J. Allen, Kilian Eichenseer, Christopher D. Dean, Sof&lt;U+00ED&gt;a Galv&lt;U+00E1&gt;n, Miranta Kouvari, et al. 2023. “Palaeoverse: A Community-Driven R Package to Support Palaeobiological Analysis.” <em>Methods in Ecology and Evolution</em>, 1–11. <a href="https://doi.org/10.1111/2041-210X.14099">https://doi.org/10.1111/2041-210X.14099</a>.</p>
</div>
<div>
<p>Legendre, Pierre, and Louis Legendre. 2012. <em>Numerical Ecology</em>. Vol. 24. Elsevier.</p>
</div>
<div>
<p>Lloyd, Graeme T. 2016. “Estimating Morphological Diversity and Tempo with Discrete Character-Taxon Matrices: Implementation, Challenges, Progress, and Future Directions.” <em>Biological Journal of the Linnean Society</em> 118 (1): 131–51.</p>
</div>
<div>
<p>———. 2018. “Journeys Through Discrete-Character Morphospace: Synthesizing Phylogeny, Tempo, and Disparity.” <em>Palaeontology</em> 61 (5): 637–45.</p>
</div>
<div>
<p>Mammola, Stefano, Carlos P Carmona, Thomas Guillerme, and Pedro Cardoso. 2021. “Concepts and Applications in Functional Diversity.” <em>Functional Ecology</em> 35 (9): 1869–85.</p>
</div>
<div>
<p>Muff, Stefanie, Erlend B Nilsen, Robert B O’Hara, and Chloé R Nater. 2022. “Rewriting Results Sections in the Language of Evidence.” <em>Trends in Ecology &amp; Evolution</em> 37 (3): 203–10.</p>
</div>
<div>
<p>Paradis, Emmanuel, and Klaus Schliep. 2019. “Ape 5.0: An Environment for Modern Phylogenetics and Evolutionary Analyses in R.” <em>Bioinformatics</em> 35: 526–28. <a href="https://doi.org/10.1093/bioinformatics/bty633">https://doi.org/10.1093/bioinformatics/bty633</a>.</p>
</div>
<div>
<p>Parins-Fukuchi, Caroline. 2018. “Use of Continuous Traits Can Improve Morphological Phylogenetics.” <em>Systematic Biology</em> 67 (2): 328–39.</p>
</div>
<div>
<p>Pennell, M. W., J. M. Eastman, G. J. Slater, J. W. Brown, J. C. Uyeda, R. G. Fitzjohn, M. E. Alfaro, and L. J. Harmon. 2014. “Geiger V2.0: An Expanded Suite of Methods for Fitting Macroevolutionary Models to Phylogenetic Trees.” <em>Bioinformatics</em> 30: 2216–8. <a href="https://doi.org/10.1093/bioinformatics/btu181">https://doi.org/10.1093/bioinformatics/btu181</a>.</p>
</div>
<div>
<p>Petchey, Owen L, and Kevin J Gaston. 2006. “Functional Diversity: Back to Basics and Looking Forward.” <em>Ecology Letters</em> 9 (6): 741–58.</p>
</div>
<div>
<p>Schlager, Stefan. 2017. “Morpho and Rvcg – Shape Analysis in R.” In <em>Statistical Shape and Deformation Analysis</em>, edited by Guoyan Zheng, Shuo Li, and Gabor Szekely, 217–56. Academic Press.</p>
</div>
<div>
<p>Shipley, Amy E, Armin Elsler, Suresh A Singh, Thomas L Stubbs, and Michael J Benton. 2024. “Locomotion and the Early Mesozoic Success of Archosauromorpha.” <em>Royal Society Open Science</em> 11 (2): 231495.</p>
</div>
<div>
<p>Stewart, Kerry, Carlos P Carmona, Chris Clements, Chris Venditti, Joseph A Tobias, and Manuela González-Suárez. 2023. “Functional Diversity Metrics Can Perform Well with Highly Incomplete Data Sets.” <em>Methods in Ecology and Evolution</em> 14 (11): 2856–72.</p>
</div>
<div>
<p>Wright, David F. 2017. “Bayesian Estimation of Fossil Phylogenies and the Evolution of Early to Middle Paleozoic Crinoids (Echinodermata).” <em>Journal of Paleontology</em> 91 (4): 799–814.</p>
</div>
</div>
</div>
</div>







<h3>References</h3>
<div id="refs" class="references hanging-indent">
<div id="ref-beck2014ancient">
<p>Beck, Robin MD, and Michael SY Lee. 2014. “Ancient Dates or Accelerated Rates? Morphological Clocks and the Antiquity of Placental Mammals.” <em>Proceedings of the Royal Society B: Biological Sciences</em> 281 (1793): 20141278.</p>
</div>
<div id="ref-bennion2023convergence">
<p>Bennion, Rebecca F, Jamie A MacLaren, Ellen J Coombs, Felix G Marx, Olivier Lambert, and Valentin Fischer. 2023. “Convergence and Constraint in the Cranial Evolution of Mosasaurid Reptiles and Early Cetaceans.” <em>Paleobiology</em> 49 (2): 215–31.</p>
</div>
<div id="ref-brennan2024evolutionary">
<p>Brennan, Ian G, David G Chapple, J Scott Keogh, and Stephen Donnellan. 2024. “Evolutionary Bursts Drive Morphological Novelty in the World’s Largest Skinks.” <em>bioRxiv</em>, 2024–06.</p>
</div>
<div id="ref-brusatte2014gradual">
<p>Brusatte, Stephen L, Graeme T Lloyd, Steve C Wang, and Mark A Norell. 2014. “Gradual Assembly of Avian Body Plan Culminated in Rapid Rates of Evolution Across the Dinosaur-Bird Transition.” <em>Current Biology</em> 24 (20): 2386–92.</p>
</div>
<div id="ref-cailliez1983analytical">
<p>Cailliez, Francis. 1983. “The Analytical Solution of the Additive Constant Problem.” <em>Psychometrika</em> 48 (2): 305–8.</p>
</div>
<div id="ref-chatar2024evolutionary">
<p>Chatar, Narimane, Margot Michaud, Davide Tamagnini, and Valentin Fischer. 2024. “Evolutionary Patterns of Cat-Like Carnivorans Unveil Drivers of the Sabertooth Morphology.” <em>Current Biology</em> 34 (11): 2460–73.</p>
</div>
<div id="ref-dushoff2019can">
<p>Dushoff, Jonathan, Morgan P Kain, and Benjamin M Bolker. 2019. “I Can See Clearly Now: Reinterpreting Statistical Significance.” <em>Methods in Ecology and Evolution</em> 10 (6): 756–59.</p>
</div>
<div id="ref-efron1994introduction">
<p>Efron, Bradley, and Robert J Tibshirani. 1994. <em>An Introduction to the Bootstrap</em>. Chapman; Hall/CRC.</p>
</div>
<div id="ref-farina2023turtle">
<p>Farina, Bruna M, Pedro L Godoy, Roger BJ Benson, Max C Langer, and Gabriel S Ferreira. 2023. “Turtle Body Size Evolution Is Determined by Lineage-Specific Specializations Rather Than Global Trends.” <em>Ecology and Evolution</em> 13 (6): e10201.</p>
</div>
<div id="ref-foffa2024comparative">
<p>Foffa, Davide, Mark T Young, and Stephen L Brusatte. 2024. “Comparative Functional Morphology Indicates Niche Partitioning Among Sympatric Marine Reptiles.” <em>Royal Society Open Science</em> 11 (5): 231951.</p>
</div>
<div id="ref-godoy2020crocodylomorph">
<p>Godoy, Pedro L. 2020. “Crocodylomorph Cranial Shape Evolution and Its Relationship with Body Size and Ecology.” <em>Journal of Evolutionary Biology</em> 33 (1): 4–21.</p>
</div>
<div id="ref-treats">
<p>Guillerme, Thomas. 2024. “Treats: A Modular R Package for Simulating Trees and Traits.” <em>Methods in Ecology and Evolution</em> 15 (4): 647–56.</p>
</div>
<div id="ref-guillerme2018time">
<p>Guillerme, Thomas, and Natalie Cooper. 2018. “Time for a Rethink: Time Sub-Sampling Methods in Disparity-Through-Time Analyses.” <em>Palaeontology</em> 61 (4): 481–93.</p>
</div>
<div id="ref-guo2024morphological">
<p>Guo, Zhen, Michael J Benton, Thomas L Stubbs, and Zhong-Qiang Chen. 2024. “Morphological Innovation Did Not Drive Diversification in Mesozoic–Cenozoic Brachiopods.” <em>Nature Ecology &amp; Evolution</em>, 1–11.</p>
</div>
<div id="ref-harmon2003tempo">
<p>Harmon, Luke J, James A Schulte, Allan Larson, and Jonathan B Losos. 2003. “Tempo and Mode of Evolutionary Radiation in Iguanian Lizards.” <em>Science</em> 301 (5635): 961–64.</p>
</div>
<div id="ref-lloyd2016estimating">
<p>Lloyd, Graeme T. 2016. “Estimating Morphological Diversity and Tempo with Discrete Character-Taxon Matrices: Implementation, Challenges, Progress, and Future Directions.” <em>Biological Journal of the Linnean Society</em> 118 (1): 131–51.</p>
</div>
<div id="ref-muff2022rewriting">
<p>Muff, Stefanie, Erlend B Nilsen, Robert B O’Hara, and Chloé R Nater. 2022. “Rewriting Results Sections in the Language of Evidence.” <em>Trends in Ecology &amp; Evolution</em> 37 (3): 203–10.</p>
</div>
<div id="ref-geiger">
<p>Pennell, M. W., J. M. Eastman, G. J. Slater, J. W. Brown, J. C. Uyeda, R. G. Fitzjohn, M. E. Alfaro, and L. J. Harmon. 2014. “Geiger V2.0: An Expanded Suite of Methods for Fitting Macroevolutionary Models to Phylogenetic Trees.” <em>Bioinformatics</em> 30: 2216–8. <a href="https://doi.org/10.1093/bioinformatics/btu181">https://doi.org/10.1093/bioinformatics/btu181</a>.</p>
</div>
<div id="ref-shipley2024locomotion">
<p>Shipley, Amy E, Armin Elsler, Suresh A Singh, Thomas L Stubbs, and Michael J Benton. 2024. “Locomotion and the Early Mesozoic Success of Archosauromorpha.” <em>Royal Society Open Science</em> 11 (2): 231495.</p>
</div>
<div id="ref-wright2017bayesian">
<p>Wright, David F. 2017. “Bayesian Estimation of Fossil Phylogenies and the Evolution of Early to Middle Paleozoic Crinoids (Echinodermata).” <em>Journal of Paleontology</em> 91 (4): 799–814.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="from-traitspaces-to-disparity.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": null,
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
